// @filoz/synapse-sdk/dist/src/sp-registry/service.js
import * as SP from '@filoz/synapse-core/sp-registry';
import { createClient, http, } from 'viem';
import { DEFAULT_CHAIN } from "../utils/constants.js";
export class SPRegistryService {
    _client;
    constructor(options) {
        this._client = options.client;
    }
    static create(options) {
        const client = createClient({
            chain: options.chain ?? DEFAULT_CHAIN,
            transport: options.transport ?? http(),
            account: options.account,
            name: 'SPRegistryService',
            key: 'sp-registry-service',
        });
        if (client.account.type === 'json-rpc' && client.transport.type !== 'custom') {
            throw new Error('Transport must be a custom transport. See https://viem.sh/docs/clients/transports/custom.');
        }
        return new SPRegistryService({ client });
    }
    async registerProvider(options) {
        const hash = await SP.registerProvider(this._client, {
            payee: options.payee,
            name: options.name,
            description: options.description,
            pdpOffering: options.pdpOffering,
            capabilities: options.capabilities,
        });
        return hash;
    }
    async updateProviderInfo(options) {
        return SP.updateProviderInfo(this._client, {
            name: options.name,
            description: options.description,
        });
    }
    async removeProvider() {
        return SP.removeProvider(this._client);
    }
    async getProvider(options) {
        try {
            return await SP.getPDPProvider(this._client, { providerId: options.providerId });
        }
        catch (error) {
            if (error instanceof Error && error.message.includes('Provider not found')) {
                return null;
            }
            throw error;
        }
    }
    async getProviderByAddress(options) {
        const providerId = await SP.getProviderIdByAddress(this._client, { providerAddress: options.address });
        if (providerId === 0n) {
            return null;
        }
        return this.getProvider({ providerId });
    }
    async getProviderIdByAddress(options) {
        return SP.getProviderIdByAddress(this._client, { providerAddress: options.address });
    }
    async getAllActiveProviders() {
        const providers = [];
        const limit = 50n;
        let offset = 0n;
        let hasMore = true;
        while (hasMore) {
            const result = await SP.getPDPProviders(this._client, {
                onlyActive: true,
                offset,
                limit,
            });
            providers.push(...result.providers);
            hasMore = result.hasMore;
            offset += limit;
        }
        return providers;
    }
    async getActiveProvidersByProductType(options) {
        const providers = [];
        const limit = 50n;
        let offset = 0n;
        let hasMore = true;
        while (hasMore) {
            const result = await SP.getProvidersByProductType(this._client, {
                productType: options.productType,
                onlyActive: true,
                offset,
                limit,
            });
            providers.push(...result.providers);
            hasMore = result.hasMore;
            offset += limit;
        }
        return providers;
    }
    async isProviderActive(options) {
        return SP.isProviderActive(this._client, { providerId: options.providerId });
    }
    async isRegisteredProvider(options) {
        return SP.isRegisteredProvider(this._client, { provider: options.address });
    }
    async getProviderCount() {
        return SP.getProviderCount(this._client);
    }
    async activeProviderCount() {
        return SP.activeProviderCount(this._client);
    }
    async addPDPProduct(options) {
        const hash = await SP.addProduct(this._client, {
            pdpOffering: options.pdpOffering,
            capabilities: options.capabilities ?? {},
        });
        return hash;
    }
    async updatePDPProduct(options) {
        const hash = await SP.updateProduct(this._client, {
            pdpOffering: options.pdpOffering,
            capabilities: options.capabilities ?? {},
        });
        return hash;
    }
    async removeProduct(options) {
        const hash = await SP.removeProduct(this._client, {
            productType: options.productType,
        });
        return hash;
    }
    async getProviders(options) {
        if (options.providerIds.length === 0) {
            return [];
        }
        return SP.getPDPProvidersByIds(this._client, {
            providerIds: options.providerIds,
        });
    }
}
