// @filoz/synapse-sdk/dist/src/filbeam/service.js
import { asChain } from '@filoz/synapse-core/chains';
import { createError } from "../utils/errors.js";
export class FilBeamService {
    _chain;
    _fetch;
    constructor(chain, fetchImpl = globalThis.fetch) {
        this._chain = asChain(chain);
        this._fetch = fetchImpl;
    }
    _getStatsBaseUrl() {
        return this._chain.testnet ? 'https://calibration.stats.filbeam.com' : 'https://stats.filbeam.com';
    }
    _validateStatsResponse(data) {
        if (typeof data !== 'object' || data === null) {
            throw createError('FilBeamService', 'validateStatsResponse', 'Response is not an object');
        }
        const response = data;
        if (typeof response.cdnEgressQuota !== 'string') {
            throw createError('FilBeamService', 'validateStatsResponse', 'cdnEgressQuota must be a string');
        }
        if (typeof response.cacheMissEgressQuota !== 'string') {
            throw createError('FilBeamService', 'validateStatsResponse', 'cacheMissEgressQuota must be a string');
        }
        return {
            cdnEgressQuota: response.cdnEgressQuota,
            cacheMissEgressQuota: response.cacheMissEgressQuota,
        };
    }
    async getDataSetStats(dataSetId) {
        const baseUrl = this._getStatsBaseUrl();
        const url = `${baseUrl}/data-set/${dataSetId}`;
        const response = await this._fetch(url, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
            },
        });
        if (response.status === 404) {
            throw createError('FilBeamService', 'getDataSetStats', `Data set not found: ${dataSetId}`);
        }
        if (response.status !== 200) {
            const errorText = await response.text().catch(() => 'Unknown error');
            throw createError('FilBeamService', 'getDataSetStats', `HTTP ${response.status} ${response.statusText}: ${errorText}`);
        }
        const data = await response.json();
        const validated = this._validateStatsResponse(data);
        return {
            cdnEgressQuota: BigInt(validated.cdnEgressQuota),
            cacheMissEgressQuota: BigInt(validated.cacheMissEgressQuota),
        };
    }
}
