// @filoz/synapse-core/dist/src/typed-data/sign-add-pieces.js
import { encodeAbiParameters, toHex, } from 'viem';
import { signTypedData } from 'viem/actions';
import { asChain } from "../chains.js";
import { randU256 } from "../utils/rand.js";
import { EIP712Types, getStorageDomain } from "./type-definitions.js";
export const signAddPiecesAbiParameters = [
    { type: 'uint256' },
    { type: 'string[][]' },
    { type: 'string[][]' },
    { type: 'bytes' },
];
export async function signAddPieces(client, options) {
    const chain = asChain(client.chain);
    const { clientDataSetId, nonce: _nonce, pieces, verifyingContract } = options;
    const nonce = _nonce ?? randU256();
    const signature = await signTypedData(client, {
        account: client.account,
        domain: getStorageDomain({ chain, verifyingContract }),
        types: EIP712Types,
        primaryType: 'AddPieces',
        message: {
            clientDataSetId,
            nonce,
            pieceData: pieces.map((piece) => {
                return {
                    data: toHex(piece.pieceCid.bytes),
                };
            }),
            pieceMetadata: pieces.map((piece, index) => ({
                pieceIndex: BigInt(index),
                metadata: piece.metadata ?? [],
            })),
        },
    });
    const metadataKV = Array.from(pieces, (piece) => piece.metadata ?? []);
    const keys = metadataKV.map((item) => item.map((item) => item.key));
    const values = metadataKV.map((item) => item.map((item) => item.value));
    const extraData = encodeAbiParameters(signAddPiecesAbiParameters, [nonce, keys, values, signature]);
    return extraData;
}
