// @filoz/synapse-core/dist/src/typed-data/sign-create-dataset.js
import { encodeAbiParameters } from 'viem';
import { signTypedData } from 'viem/actions';
import { asChain } from "../chains.js";
import { randU256 } from "../utils/rand.js";
import { EIP712Types, getStorageDomain } from "./type-definitions.js";
export const signCreateDataSetAbiParameters = [
    { type: 'address' },
    { type: 'uint256' },
    { type: 'string[]' },
    { type: 'string[]' },
    { type: 'bytes' },
];
export async function signCreateDataSet(client, options) {
    const chain = asChain(client.chain);
    const metadata = options.metadata ?? [];
    const clientDataSetId = options.clientDataSetId ?? randU256();
    const signature = await signTypedData(client, {
        account: client.account,
        domain: getStorageDomain({ chain }),
        types: EIP712Types,
        primaryType: 'CreateDataSet',
        message: {
            clientDataSetId,
            payee: options.payee,
            metadata,
        },
    });
    const keys = metadata.map((item) => item.key);
    const values = metadata.map((item) => item.value);
    const payer = options.payer ?? client.account.address;
    const extraData = encodeAbiParameters(signCreateDataSetAbiParameters, [
        payer,
        clientDataSetId,
        keys,
        values,
        signature,
    ]);
    return extraData;
}
