// @filoz/synapse-core/dist/src/typed-data/type-definitions.js
export const EIP712Types = {
    MetadataEntry: [
        { name: 'key', type: 'string' },
        { name: 'value', type: 'string' },
    ],
    CreateDataSet: [
        { name: 'clientDataSetId', type: 'uint256' },
        { name: 'payee', type: 'address' },
        { name: 'metadata', type: 'MetadataEntry[]' },
    ],
    Cid: [{ name: 'data', type: 'bytes' }],
    PieceMetadata: [
        { name: 'pieceIndex', type: 'uint256' },
        { name: 'metadata', type: 'MetadataEntry[]' },
    ],
    AddPieces: [
        { name: 'clientDataSetId', type: 'uint256' },
        { name: 'nonce', type: 'uint256' },
        { name: 'pieceData', type: 'Cid[]' },
        { name: 'pieceMetadata', type: 'PieceMetadata[]' },
    ],
    SchedulePieceRemovals: [
        { name: 'clientDataSetId', type: 'uint256' },
        { name: 'pieceIds', type: 'uint256[]' },
    ],
    DeleteDataSet: [{ name: 'clientDataSetId', type: 'uint256' }],
    Permit: [
        { name: 'owner', type: 'address' },
        { name: 'spender', type: 'address' },
        { name: 'value', type: 'uint256' },
        { name: 'nonce', type: 'uint256' },
        { name: 'deadline', type: 'uint256' },
    ],
};
export function getStorageDomain(options) {
    return {
        name: 'FilecoinWarmStorageService',
        version: '1',
        chainId: options.chain.id,
        verifyingContract: options.verifyingContract ?? options.chain.contracts.fwss.address,
    };
}
