// @filoz/synapse-core/dist/src/pdp-verifier/get-active-pieces.js
import { readContract } from 'viem/actions';
import { asChain } from "../chains.js";
import { hexToPieceCID } from "../piece.js";
export async function getActivePieces(client, options) {
    const data = await readContract(client, getActivePiecesCall({
        chain: client.chain,
        dataSetId: options.dataSetId,
        offset: options.offset,
        limit: options.limit,
        contractAddress: options.contractAddress,
    }));
    return parseActivePieces(data);
}
export function getActivePiecesCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.pdp.abi,
        address: options.contractAddress ?? chain.contracts.pdp.address,
        functionName: 'getActivePieces',
        args: [options.dataSetId, options.offset ?? 0n, options.limit ?? 100n],
    };
}
export function parseActivePieces(data) {
    return {
        pieces: data[0].map((piece, index) => ({
            cid: hexToPieceCID(piece.data),
            id: data[1][index],
        })),
        hasMore: data[2],
    };
}
