// @filoz/synapse-core/dist/src/pdp-verifier/get-pieces.js
import { multicall } from 'viem/actions';
import { asChain } from "../chains.js";
import { hexToPieceCID } from "../piece.js";
import { metadataArrayToObject } from "../utils/metadata.js";
import { createPieceUrl } from "../utils/piece-url.js";
import { getAllPieceMetadataCall } from "../warm-storage/get-all-piece-metadata.js";
import { getActivePiecesCall } from "./get-active-pieces.js";
import { getScheduledRemovalsCall } from "./get-scheduled-removals.js";
export async function getPieces(client, options) {
    const chain = asChain(client.chain);
    const address = options.address;
    const serviceURL = options.dataSet.provider.pdp.serviceURL;
    const [activePiecesResult, removalsResult] = await multicall(client, {
        contracts: [
            getActivePiecesCall({
                chain: client.chain,
                dataSetId: options.dataSet.dataSetId,
                offset: options.offset,
                limit: options.limit,
                contractAddress: options.contractAddress,
            }),
            getScheduledRemovalsCall({
                chain: client.chain,
                dataSetId: options.dataSet.dataSetId,
                contractAddress: options.contractAddress,
            }),
        ],
        allowFailure: false,
    });
    const removals = Array.from(new Set(removalsResult));
    return {
        hasMore: activePiecesResult[2],
        pieces: activePiecesResult[0]
            .map((piece, index) => {
            const cid = hexToPieceCID(piece.data);
            return {
                cid,
                id: activePiecesResult[1][index],
                url: createPieceUrl({
                    cid: cid.toString(),
                    cdn: options.dataSet.cdn,
                    address,
                    chain,
                    serviceURL,
                }),
            };
        })
            .filter((piece) => !removals.includes(piece.id)),
    };
}
export async function getPiecesWithMetadata(client, options) {
    const pieces = await getPieces(client, options);
    const metadata = await multicall(client, {
        allowFailure: false,
        contracts: pieces.pieces.map((piece) => getAllPieceMetadataCall({
            chain: client.chain,
            dataSetId: options.dataSet.dataSetId,
            pieceId: piece.id,
            contractAddress: options.contractAddress,
        })),
    });
    return {
        pieces: pieces.pieces.map((piece, index) => ({
            ...piece,
            metadata: metadataArrayToObject(metadata[index]),
        })),
        hasMore: pieces.hasMore,
    };
}
