// @filoz/synapse-core/dist/src/auction/auction.js
import { readContract } from 'viem/actions';
import { getChain } from "../chains.js";
import { PRB_ONE, prbDiv, prbExp2 } from "./prb.js";
export async function auctionInfo(client, token) {
    const chain = getChain(client.chain.id);
    const [startPrice, startTime] = await readContract(client, {
        address: chain.contracts.filecoinPay.address,
        abi: chain.contracts.filecoinPay.abi,
        functionName: 'auctionInfo',
        args: [token],
    });
    return {
        startPrice,
        startTime,
        token,
    };
}
export async function auctionFunds(client, token) {
    const chain = getChain(client.chain.id);
    const [funds] = await readContract(client, {
        address: chain.contracts.filecoinPay.address,
        abi: chain.contracts.filecoinPay.abi,
        functionName: 'accounts',
        args: [token, chain.contracts.filecoinPay.address],
    });
    return funds;
}
export const HALVING_SECONDS = 7n * 12n * 60n * 60n;
const MAX_DELAY = HALVING_SECONDS * 192n;
export function auctionPriceAt(auction, timestamp) {
    const elapsed = timestamp - auction.startTime;
    if (elapsed <= 0) {
        return auction.startPrice;
    }
    if (elapsed >= MAX_DELAY) {
        return 0n;
    }
    return prbDiv(auction.startPrice, prbExp2((elapsed * PRB_ONE) / HALVING_SECONDS));
}
