// @filoz/synapse-core/dist/src/sp-registry/add-product.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
import { encodePDPCapabilities } from "../utils/pdp-capabilities.js";
export async function addProduct(client, options) {
    const [capabilityKeys, capabilityValues] = encodePDPCapabilities(options.pdpOffering, options.capabilities);
    const { request } = await simulateContract(client, addProductCall({
        chain: client.chain,
        productType: options.productType,
        capabilityKeys,
        capabilityValues,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function addProductSync(client, options) {
    const hash = await addProduct(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractAddProductEvent(receipt.logs);
    return { receipt, event };
}
export function addProductCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.serviceProviderRegistry.abi,
        address: options.contractAddress ?? chain.contracts.serviceProviderRegistry.address,
        functionName: 'addProduct',
        args: [options.productType ?? 0, options.capabilityKeys, options.capabilityValues],
    };
}
export function extractAddProductEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.serviceProviderRegistry,
        logs,
        eventName: 'ProductAdded',
        strict: true,
    });
    if (!log)
        throw new Error('`ProductAdded` event not found.');
    return log;
}
