// @filoz/synapse-core/dist/src/sp-registry/get-pdp-providers.js
import { multicall, readContract } from 'viem/actions';
import { asChain } from "../chains.js";
import { getApprovedProvidersCall } from "../warm-storage/get-approved-providers.js";
import { getPDPProviderCall, parsePDPProvider } from "./get-pdp-provider.js";
import { PRODUCTS } from "./types.js";
export async function getPDPProviders(client, options = {}) {
    const data = await readContract(client, getPDPProvidersCall({
        chain: client.chain,
        onlyActive: options.onlyActive,
        offset: options.offset,
        limit: options.limit,
        contractAddress: options.contractAddress,
    }));
    return parsePDPProviders(data);
}
export function getPDPProvidersCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.serviceProviderRegistry.abi,
        address: options.contractAddress ?? chain.contracts.serviceProviderRegistry.address,
        functionName: 'getProvidersByProductType',
        args: [PRODUCTS.PDP, options.onlyActive ?? true, options.offset ?? 0n, options.limit ?? 50n],
    };
}
export function parsePDPProviders(data) {
    return {
        providers: data.providers.map(parsePDPProvider),
        hasMore: data.hasMore,
    };
}
export async function getApprovedPDPProviders(client, options = {}) {
    const [pdpProviders, approvedProviders] = await multicall(client, {
        allowFailure: false,
        contracts: [
            getPDPProvidersCall({
                chain: client.chain,
                onlyActive: true,
                offset: 0n,
                limit: 100n,
                contractAddress: options.contractAddress,
            }),
            getApprovedProvidersCall({
                chain: client.chain,
            }),
        ],
    });
    const providers = [];
    for (const provider of pdpProviders.providers) {
        if (approvedProviders.includes(provider.providerId)) {
            providers.push(provider);
        }
    }
    return parsePDPProviders({
        providers,
        hasMore: false,
    }).providers;
}
export async function getPDPProvidersByIds(client, options) {
    const result = await multicall(client, {
        allowFailure: true,
        contracts: options.providerIds.map((providerId) => getPDPProviderCall({
            chain: client.chain,
            providerId,
            contractAddress: options.contractAddress,
        })),
    });
    return parsePDPProviders({
        providers: result.filter((result) => result.status === 'success').map((result) => result.result),
        hasMore: false,
    }).providers;
}
