// @filoz/synapse-core/dist/src/sp-registry/register-provider.js
import { parseEventLogs } from 'viem';
import { readContract, simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
import { encodePDPCapabilities } from "../utils/pdp-capabilities.js";
export async function registerProvider(client, options) {
    const chain = asChain(client.chain);
    const contractAddress = options.contractAddress ?? chain.contracts.serviceProviderRegistry.address;
    let registrationFee;
    if (options.value !== undefined) {
        registrationFee = options.value;
    }
    else {
        registrationFee = await readContract(client, {
            abi: chain.contracts.serviceProviderRegistry.abi,
            address: contractAddress,
            functionName: 'REGISTRATION_FEE',
        });
    }
    const { request } = await simulateContract(client, registerProviderCall({
        chain: client.chain,
        payee: options.payee,
        name: options.name,
        description: options.description,
        productType: options.productType,
        pdpOffering: options.pdpOffering,
        capabilities: options.capabilities,
        value: registrationFee,
        contractAddress,
    }));
    return writeContract(client, request);
}
export async function registerProviderSync(client, options) {
    const hash = await registerProvider(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractRegisterProviderEvent(receipt.logs);
    return { receipt, event };
}
export function registerProviderCall(options) {
    const chain = asChain(options.chain);
    const [capabilityKeys, capabilityValues] = encodePDPCapabilities(options.pdpOffering, options.capabilities);
    return {
        abi: chain.contracts.serviceProviderRegistry.abi,
        address: options.contractAddress ?? chain.contracts.serviceProviderRegistry.address,
        functionName: 'registerProvider',
        args: [
            options.payee,
            options.name,
            options.description,
            options.productType ?? 0,
            capabilityKeys,
            capabilityValues,
        ],
        value: options.value,
    };
}
export function extractRegisterProviderEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.serviceProviderRegistry,
        logs,
        eventName: 'ProviderRegistered',
        strict: true,
    });
    if (!log)
        throw new Error('`ProviderRegistered` event not found.');
    return log;
}
