// @filoz/synapse-core/dist/src/sp-registry/remove-provider.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
export async function removeProvider(client, options = {}) {
    const { request } = await simulateContract(client, removeProviderCall({
        chain: client.chain,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function removeProviderSync(client, options = {}) {
    const hash = await removeProvider(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractRemoveProviderEvent(receipt.logs);
    return { receipt, event };
}
export function removeProviderCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.serviceProviderRegistry.abi,
        address: options.contractAddress ?? chain.contracts.serviceProviderRegistry.address,
        functionName: 'removeProvider',
        args: [],
    };
}
export function extractRemoveProviderEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.serviceProviderRegistry,
        logs,
        eventName: 'ProviderRemoved',
        strict: true,
    });
    if (!log)
        throw new Error('`ProviderRemoved` event not found.');
    return log;
}
