// @filoz/synapse-core/dist/src/sp-registry/remove-product.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
export async function removeProduct(client, options = {}) {
    const { request } = await simulateContract(client, removeProductCall({
        chain: client.chain,
        productType: options.productType,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function removeProductSync(client, options = {}) {
    const hash = await removeProduct(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractRemoveProductEvent(receipt.logs);
    return { receipt, event };
}
export function removeProductCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.serviceProviderRegistry.abi,
        address: options.contractAddress ?? chain.contracts.serviceProviderRegistry.address,
        functionName: 'removeProduct',
        args: [options.productType ?? 0],
    };
}
export function extractRemoveProductEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.serviceProviderRegistry,
        logs,
        eventName: 'ProductRemoved',
        strict: true,
    });
    if (!log)
        throw new Error('`ProductRemoved` event not found.');
    return log;
}
