// @filoz/synapse-core/dist/src/sp-registry/update-provider-info.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
export async function updateProviderInfo(client, options) {
    const chain = asChain(client.chain);
    const contractAddress = options.contractAddress ?? chain.contracts.serviceProviderRegistry.address;
    const { request } = await simulateContract(client, updateProviderInfoCall({
        chain: client.chain,
        name: options.name,
        description: options.description,
        contractAddress,
    }));
    return writeContract(client, request);
}
export async function updateProviderInfoSync(client, options) {
    const hash = await updateProviderInfo(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractUpdateProviderInfoEvent(receipt.logs);
    return { receipt, event };
}
export function updateProviderInfoCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.serviceProviderRegistry.abi,
        address: options.contractAddress ?? chain.contracts.serviceProviderRegistry.address,
        functionName: 'updateProviderInfo',
        args: [options.name, options.description],
    };
}
export function extractUpdateProviderInfoEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.serviceProviderRegistry,
        logs,
        eventName: 'ProviderInfoUpdated',
        strict: true,
    });
    if (!log)
        throw new Error('`ProviderInfoUpdated` event not found.');
    return log;
}
