// @filoz/synapse-core/dist/src/erc20/approve.js
import { erc20Abi, parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import { asChain } from "../chains.js";
import { AllowanceAmountError } from "../errors/erc20.js";
export async function approve(client, options) {
    const { request } = await simulateContract(client, approveCall({
        chain: client.chain,
        token: options.token,
        spender: options.spender,
        amount: options.amount,
    }));
    return writeContract(client, request);
}
export async function approveSync(client, options) {
    const hash = await approve(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractApproveEvent(receipt.logs);
    return { receipt, event };
}
export function approveCall(options) {
    const chain = asChain(options.chain);
    const token = options.token ?? chain.contracts.usdfc.address;
    const spender = options.spender ?? chain.contracts.filecoinPay.address;
    if (options.amount < 0n) {
        throw new AllowanceAmountError(options.amount);
    }
    return {
        abi: erc20Abi,
        address: token,
        functionName: 'approve',
        args: [spender, options.amount],
    };
}
export function extractApproveEvent(logs) {
    const [log] = parseEventLogs({
        abi: erc20Abi,
        logs,
        eventName: 'Approval',
        strict: true,
    });
    if (!log)
        throw new Error('`Approval` event not found.');
    return log;
}
