// @filoz/synapse-core/dist/src/erc20/index.js
import { erc20Abi } from 'viem';
import { multicall } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
export async function balance(client, options) {
    const chain = asChain(client.chain);
    const token = options.token ?? chain.contracts.usdfc.address;
    const result = await multicall(client, {
        allowFailure: false,
        contracts: [
            {
                address: token,
                abi: erc20Abi,
                functionName: 'balanceOf',
                args: [options.address],
            },
            {
                address: token,
                abi: erc20Abi,
                functionName: 'decimals',
            },
            {
                address: token,
                abi: erc20Abi,
                functionName: 'symbol',
            },
            {
                address: token,
                abi: erc20Abi,
                functionName: 'allowance',
                args: [options.address, options.spender ?? chain.contracts.filecoinPay.address],
            },
        ],
    });
    return {
        value: result[0],
        decimals: result[1],
        symbol: result[2],
        allowance: result[3],
    };
}
export async function balanceForPermit(client, options) {
    const chain = asChain(client.chain);
    const token = options.token ?? chain.contracts.usdfc.address;
    const result = await multicall(client, {
        allowFailure: false,
        contracts: [
            {
                address: token,
                abi: Abis.erc20WithPermit,
                functionName: 'balanceOf',
                args: [options.address],
            },
            {
                address: token,
                abi: Abis.erc20WithPermit,
                functionName: 'name',
            },
            {
                address: token,
                abi: Abis.erc20WithPermit,
                functionName: 'nonces',
                args: [options.address],
            },
            {
                address: token,
                abi: Abis.erc20WithPermit,
                functionName: 'version',
            },
        ],
    });
    return {
        value: result[0],
        name: result[1],
        nonce: result[2],
        version: result[3],
    };
}
export * from "./approve.js";
