// @filoz/synapse-core/dist/src/errors/base.js
import * as z from 'zod';
const symbol = Symbol.for('synapse-error');
export function isSynapseError(value) {
    return value instanceof Error && symbol in value;
}
export class SynapseError extends Error {
    [symbol] = true;
    name = 'SynapseError';
    cause;
    details;
    shortMessage;
    constructor(message, options) {
        const details = options?.details
            ? options.details
            : options?.cause instanceof Error
                ? options.cause.message
                : undefined;
        const msg = [
            message || 'An error occurred.',
            ...(details ? [''] : []),
            ...(details ? [`Details: ${details}`] : []),
        ].join('\n');
        super(msg, options);
        this.cause = options?.cause ?? undefined;
        this.details = details ?? undefined;
        this.shortMessage = message;
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'SynapseError';
    }
}
export class ZodValidationError extends SynapseError {
    name = 'ZodValidationError';
    constructor(zodError, message = 'Validation failed.') {
        super(message, {
            cause: zodError,
            details: z.prettifyError(zodError),
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'ZodValidationError';
    }
}
export class ValidationError extends SynapseError {
    name = 'ValidationError';
    static is(value) {
        return isSynapseError(value) && value.name === 'ValidationError';
    }
}
