// @filoz/synapse-core/dist/src/errors/pay.js
import { isSynapseError, SynapseError } from "./base.js";
export class InsufficientBalanceError extends SynapseError {
    name = 'InsufficientBalanceError';
    constructor(balance, required) {
        super(`Insufficient balance.`, {
            details: `Balance: ${balance}, Required: ${required}`,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'InsufficientBalanceError';
    }
}
export class InsufficientAllowanceError extends SynapseError {
    name = 'InsufficientAllowanceError';
    constructor(allowance, required) {
        super(`Insufficient allowance.`, {
            details: `Allowance: ${allowance}, Required: ${required}`,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'InsufficientAllowanceError';
    }
}
export class DepositAmountError extends SynapseError {
    name = 'DepositAmountError';
    constructor(amount) {
        super(`Deposit amount must be greater than 0.`, {
            details: `Amount: ${amount}`,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'DepositAmountError';
    }
}
export class WithdrawAmountError extends SynapseError {
    name = 'WithdrawAmountError';
    constructor(amount) {
        super(`Withdraw amount must be greater than 0.`, {
            details: `Amount: ${amount}`,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'WithdrawAmountError';
    }
}
export class InsufficientAvailableFundsError extends SynapseError {
    name = 'InsufficientAvailableFundsError';
    constructor(availableFunds, required) {
        super(`Insufficient available funds.`, {
            details: `Available funds: ${availableFunds}, Required: ${required}`,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'InsufficientAvailableFundsError';
    }
}
