// @filoz/synapse-core/dist/src/errors/pdp.js
import { SIZE_CONSTANTS } from "../utils/constants.js";
import { decodePDPError } from "../utils/decode-pdp-errors.js";
import { isSynapseError, SynapseError } from "./base.js";
export class LocationHeaderError extends SynapseError {
    name = 'LocationHeaderError';
    constructor(location) {
        super(`Location header format is invalid: ${location ?? '<none>'}`);
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'LocationHeaderError';
    }
}
export class CreateDataSetError extends SynapseError {
    name = 'CreateDataSetError';
    constructor(error) {
        const decodedError = decodePDPError(error);
        super(`Failed to create data set.`, {
            details: decodedError,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'CreateDataSetError';
    }
}
export class WaitForCreateDataSetError extends SynapseError {
    name = 'WaitForCreateDataSetError';
    constructor(error) {
        const decodedError = decodePDPError(error);
        super(`Failed to wait for data set creation.`, {
            details: decodedError,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'WaitForCreateDataSetError';
    }
}
export class WaitForCreateDataSetRejectedError extends SynapseError {
    name = 'WaitForCreateDataSetRejectedError';
    response;
    constructor(error) {
        super(`Data set creation request rejected.`, {
            details: `Tx hash: ${error.createMessageHash}`,
        });
        this.response = error;
    }
}
export class GetDataSetError extends SynapseError {
    name = 'GetDataSetError';
    constructor(error) {
        super(error ? 'Failed to get data set.' : 'Data set not found.', {
            details: error ? decodePDPError(error) : undefined,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'GetDataSetError';
    }
}
export class PostPieceError extends SynapseError {
    name = 'PostPieceError';
    constructor(error) {
        const decodedError = decodePDPError(error);
        super(`Failed to create upload session.`, {
            details: decodedError,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'PostPieceError';
    }
}
export class UploadPieceError extends SynapseError {
    name = 'UploadPieceError';
    constructor(error) {
        const decodedError = decodePDPError(error);
        super(`Failed to upload piece.`, {
            details: decodedError,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'UploadPieceError';
    }
}
export class FindPieceError extends SynapseError {
    name = 'FindPieceError';
    constructor(error) {
        const decodedError = decodePDPError(error);
        super(`Failed to find piece.`, {
            details: decodedError,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'FindPieceError';
    }
}
export class AddPiecesError extends SynapseError {
    name = 'AddPiecesError';
    constructor(error) {
        const decodedError = decodePDPError(error);
        super(`Failed to add pieces.`, {
            details: decodedError,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'AddPiecesError';
    }
}
export class WaitForAddPiecesError extends SynapseError {
    name = 'WaitForAddPiecesError';
    constructor(error) {
        const decodedError = decodePDPError(error);
        super(`Failed to wait for add pieces.`, {
            details: decodedError,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'WaitForAddPiecesError';
    }
}
export class WaitForAddPiecesRejectedError extends SynapseError {
    name = 'WaitForAddPiecesRejectedError';
    response;
    constructor(error) {
        super(`Add pieces request rejected.`, {
            details: `Tx hash: ${error.txHash}, Data set ID: ${error.dataSetId}, Piece count: ${error.pieceCount}`,
        });
        this.response = error;
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'WaitForAddPiecesRejectedError';
    }
}
export class DeletePieceError extends SynapseError {
    name = 'DeletePieceError';
    constructor(error) {
        const decodedError = decodePDPError(error);
        super(`Failed to delete piece.`, {
            details: decodedError,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'DeletePieceError';
    }
}
export class InvalidUploadSizeError extends SynapseError {
    name = 'InvalidUploadSizeError';
    constructor(size) {
        super(`Invalid upload size.`, {
            details: `Size ${size} bytes is below minimum allowed size of ${SIZE_CONSTANTS.MIN_UPLOAD_SIZE} bytes or exceeds maximum allowed size of ${SIZE_CONSTANTS.MAX_UPLOAD_SIZE} bytes (1 GiB with fr32 expansion)`,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'InvalidUploadSizeError';
    }
}
export class DownloadPieceError extends SynapseError {
    name = 'DownloadPieceError';
    constructor(error) {
        super(`Failed to download piece.`, {
            details: error,
        });
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'DownloadPieceError';
    }
}
