// @filoz/synapse-sdk/dist/src/storage/context.js
import { asChain } from '@filoz/synapse-core/chains';
import { getProviderIds } from '@filoz/synapse-core/endorsements';
import * as PDPVerifier from '@filoz/synapse-core/pdp-verifier';
import { asPieceCID } from '@filoz/synapse-core/piece';
import * as SP from '@filoz/synapse-core/sp';
import { schedulePieceDeletion } from '@filoz/synapse-core/sp';
import { calculateLastProofDate, createPieceUrlPDP, epochToDate, pieceMetadataObjectToEntry, randIndex, timeUntilEpoch, } from '@filoz/synapse-core/utils';
import { getBlockNumber } from 'viem/actions';
import { SPRegistryService } from "../sp-registry/index.js";
import { createError, METADATA_KEYS, SIZE_CONSTANTS } from "../utils/index.js";
import { combineMetadata, metadataMatches } from "../utils/metadata.js";
const NO_REMAINING_PROVIDERS_ERROR_MESSAGE = 'No approved service providers available';
export class StorageContext {
    _client;
    _chain;
    _synapse;
    _provider;
    _pdpEndpoint;
    _warmStorageService;
    _withCDN;
    _uploadBatchSize;
    _dataSetId;
    _clientDataSetId;
    _dataSetMetadata;
    _pendingPieces = [];
    _isProcessing = false;
    _activeUploads = new Set();
    _uploadBatchWaitTimeout = 15000;
    serviceProvider;
    get withCDN() {
        return this._withCDN;
    }
    get provider() {
        return this._provider;
    }
    get dataSetMetadata() {
        return this._dataSetMetadata;
    }
    get dataSetId() {
        return this._dataSetId;
    }
    async getClientDataSetId() {
        if (this._clientDataSetId !== undefined) {
            return this._clientDataSetId;
        }
        if (this.dataSetId == null) {
            throw createError('StorageContext', 'getClientDataSetId', 'Data set not found');
        }
        const dataSetInfo = await this._warmStorageService.getDataSet({ dataSetId: this.dataSetId });
        if (dataSetInfo == null) {
            throw createError('StorageContext', 'getClientDataSetId', 'Data set not found');
        }
        this._clientDataSetId = dataSetInfo.clientDataSetId;
        return this._clientDataSetId;
    }
    static validateRawSize(options) {
        const { sizeBytes, context } = options;
        if (sizeBytes < SIZE_CONSTANTS.MIN_UPLOAD_SIZE) {
            throw createError('StorageContext', context, `Data size ${sizeBytes} bytes is below minimum allowed size of ${SIZE_CONSTANTS.MIN_UPLOAD_SIZE} bytes`);
        }
        if (sizeBytes > SIZE_CONSTANTS.MAX_UPLOAD_SIZE) {
            throw createError('StorageContext', context, `Data size ${sizeBytes} bytes exceeds maximum allowed size of ${SIZE_CONSTANTS.MAX_UPLOAD_SIZE} bytes (${Math.floor(SIZE_CONSTANTS.MAX_UPLOAD_SIZE / 1024 / 1024)} MiB)`);
        }
    }
    constructor(options) {
        this._client = options.synapse.client;
        this._chain = asChain(this._client.chain);
        this._synapse = options.synapse;
        this._provider = options.provider;
        this._withCDN = options.options.withCDN ?? false;
        this._warmStorageService = options.warmStorageService;
        this._uploadBatchSize = Math.max(1, options.options.uploadBatchSize ?? SIZE_CONSTANTS.DEFAULT_UPLOAD_BATCH_SIZE);
        this._dataSetMetadata = options.dataSetMetadata;
        this._dataSetId = options.dataSetId;
        this.serviceProvider = options.provider.serviceProvider;
        this._pdpEndpoint = options.provider.pdp.serviceURL;
    }
    static async createContexts(options) {
        const count = options?.count ?? 2;
        const resolutions = [];
        const clientAddress = options.synapse.client.account.address;
        const spRegistry = new SPRegistryService({ client: options.synapse.client });
        if (options.dataSetIds) {
            const selections = [];
            for (const dataSetId of new Set(options.dataSetIds)) {
                selections.push(StorageContext.resolveByDataSetId(dataSetId, options.warmStorageService, spRegistry, clientAddress, {
                    withCDN: options.withCDN,
                    metadata: options.metadata,
                }));
                if (selections.length >= count) {
                    break;
                }
            }
            resolutions.push(...(await Promise.all(selections)));
        }
        const resolvedProviderIds = resolutions.map((resolution) => resolution.provider.id);
        if (resolutions.length < count) {
            if (options.providerIds) {
                const selections = [];
                for (const providerId of [...new Set(options.providerIds)].filter((providerId) => !resolvedProviderIds.includes(providerId))) {
                    selections.push(StorageContext.resolveByProviderId(clientAddress, providerId, options.metadata ?? {}, options.warmStorageService, spRegistry, options.forceCreateDataSets));
                    resolvedProviderIds.push(providerId);
                    if (selections.length + resolutions.length >= count) {
                        break;
                    }
                }
                resolutions.push(...(await Promise.all(selections)));
            }
        }
        if (resolutions.length < count) {
            const excludeProviderIds = [...(options.excludeProviderIds ?? []), ...resolvedProviderIds];
            for (let i = resolutions.length; i < count; i++) {
                try {
                    const resolution = await StorageContext.smartSelectProvider(clientAddress, options.metadata ?? {}, options.warmStorageService, spRegistry, excludeProviderIds, resolutions.length === 0 ? await getProviderIds(options.synapse.client) : new Set(), options.forceCreateDataSets ?? false);
                    excludeProviderIds.push(resolution.provider.id);
                    resolutions.push(resolution);
                }
                catch (error) {
                    if (error instanceof Error && error.message.includes(NO_REMAINING_PROVIDERS_ERROR_MESSAGE)) {
                        break;
                    }
                    throw error;
                }
            }
        }
        return await Promise.all(resolutions.map(async (resolution) => await StorageContext.createWithSelectedProvider(resolution, options.synapse, options.warmStorageService, options)));
    }
    static async create(options) {
        const spRegistry = new SPRegistryService({ client: options.synapse.client });
        const resolution = await StorageContext.resolveProviderAndDataSet(options.synapse, options.warmStorageService, spRegistry, options);
        return await StorageContext.createWithSelectedProvider(resolution, options.synapse, options.warmStorageService, options);
    }
    static async createWithSelectedProvider(resolution, synapse, warmStorageService, options = {}) {
        try {
            options.callbacks?.onProviderSelected?.(resolution.provider);
        }
        catch (error) {
            console.error('Error in onProviderSelected callback:', error);
        }
        if (resolution.dataSetId !== -1n) {
            options.callbacks?.onDataSetResolved?.({
                isExisting: resolution.dataSetId !== -1n,
                dataSetId: resolution.dataSetId,
                provider: resolution.provider,
            });
        }
        return new StorageContext({
            synapse,
            warmStorageService,
            provider: resolution.provider,
            dataSetId: resolution.dataSetId === -1n ? undefined : resolution.dataSetId,
            options,
            dataSetMetadata: resolution.dataSetMetadata,
        });
    }
    static async resolveProviderAndDataSet(synapse, warmStorageService, spRegistry, options) {
        const clientAddress = synapse.client.account.address;
        if (options.dataSetId != null && options.forceCreateDataSet !== true) {
            return await StorageContext.resolveByDataSetId(options.dataSetId, warmStorageService, spRegistry, clientAddress, options);
        }
        const requestedMetadata = combineMetadata(options.metadata, options.withCDN);
        if (options.providerId != null) {
            return await StorageContext.resolveByProviderId(clientAddress, options.providerId, requestedMetadata, warmStorageService, spRegistry, options.forceCreateDataSet);
        }
        if (options.providerAddress != null) {
            return await StorageContext.resolveByProviderAddress(options.providerAddress, warmStorageService, spRegistry, clientAddress, requestedMetadata, options.forceCreateDataSet);
        }
        return await StorageContext.smartSelectProvider(clientAddress, requestedMetadata, warmStorageService, spRegistry, options.excludeProviderIds ?? [], new Set(), options.forceCreateDataSet ?? false);
    }
    static async resolveByDataSetId(dataSetId, warmStorageService, spRegistry, clientAddress, options) {
        const [dataSetInfo, dataSetMetadata] = await Promise.all([
            warmStorageService.getDataSet({ dataSetId }).then(async (dataSetInfo) => {
                if (dataSetInfo == null) {
                    return null;
                }
                await StorageContext.validateDataSetConsistency(dataSetInfo, options, spRegistry);
                return dataSetInfo;
            }),
            warmStorageService.getDataSetMetadata({ dataSetId }),
            warmStorageService.validateDataSet({ dataSetId }),
        ]);
        if (dataSetInfo == null) {
            throw createError('StorageContext', 'resolveByDataSetId', `Data set ${dataSetId} does not exist`);
        }
        if (dataSetInfo.payer.toLowerCase() !== clientAddress.toLowerCase()) {
            throw createError('StorageContext', 'resolveByDataSetId', `Data set ${dataSetId} is not owned by ${clientAddress} (owned by ${dataSetInfo.payer})`);
        }
        const provider = await spRegistry.getProvider({ providerId: dataSetInfo.providerId });
        if (provider == null) {
            throw createError('StorageContext', 'resolveByDataSetId', `Provider ID ${dataSetInfo.providerId} for data set ${dataSetId} not found in registry`);
        }
        const withCDN = dataSetInfo.cdnRailId > 0 && METADATA_KEYS.WITH_CDN in dataSetMetadata;
        if (options.withCDN != null && withCDN !== options.withCDN) {
            throw createError('StorageContext', 'resolveByDataSetId', `Data set ${dataSetId} has CDN ${withCDN ? 'enabled' : 'disabled'}, ` +
                `but requested ${options.withCDN ? 'enabled' : 'disabled'}`);
        }
        return {
            provider,
            dataSetId,
            isExisting: true,
            dataSetMetadata,
        };
    }
    static async validateDataSetConsistency(dataSet, options, spRegistry) {
        if (options.providerId != null) {
            if (dataSet.providerId !== options.providerId) {
                throw createError('StorageContext', 'validateDataSetConsistency', `Data set belongs to provider ID ${dataSet.providerId}, but provider ID ${options.providerId} was requested`);
            }
        }
        if (options.providerAddress != null) {
            const actualProvider = await spRegistry.getProvider({ providerId: dataSet.providerId });
            if (actualProvider == null ||
                actualProvider.serviceProvider.toLowerCase() !== options.providerAddress.toLowerCase()) {
                throw createError('StorageContext', 'validateDataSetConsistency', `Data set belongs to provider ${actualProvider?.serviceProvider ?? 'unknown'}, but provider ${options.providerAddress} was requested`);
            }
        }
    }
    static async resolveByProviderId(clientAddress, providerId, requestedMetadata, warmStorageService, spRegistry, forceCreateDataSet) {
        const [provider, dataSets] = await Promise.all([
            spRegistry.getProvider({ providerId }),
            forceCreateDataSet ? Promise.resolve([]) : warmStorageService.getClientDataSets({ address: clientAddress }),
        ]);
        if (provider == null) {
            throw createError('StorageContext', 'resolveByProviderId', `Provider ID ${providerId} not found in registry`);
        }
        if (forceCreateDataSet === true) {
            return {
                provider,
                dataSetId: -1n,
                isExisting: false,
                dataSetMetadata: requestedMetadata,
            };
        }
        const providerDataSets = dataSets.filter((dataSet) => dataSet.dataSetId && dataSet.providerId === provider.id && dataSet.pdpEndEpoch === 0n);
        const sortedDataSets = providerDataSets.sort((a, b) => {
            return Number(a.dataSetId) - Number(b.dataSetId);
        });
        const MIN_BATCH_SIZE = 50;
        const MAX_BATCH_SIZE = 200;
        const BATCH_SIZE = Math.min(MAX_BATCH_SIZE, Math.max(MIN_BATCH_SIZE, Math.ceil(sortedDataSets.length / 3), 1));
        let selectedDataSet = null;
        for (let i = 0; i < sortedDataSets.length; i += BATCH_SIZE) {
            const batchResults = await Promise.all(sortedDataSets.slice(i, i + BATCH_SIZE).map(async (dataSet) => {
                const dataSetId = dataSet.dataSetId;
                try {
                    const [dataSetMetadata, activePieceCount] = await Promise.all([
                        warmStorageService.getDataSetMetadata({ dataSetId }),
                        warmStorageService.getActivePieceCount({ dataSetId }),
                        warmStorageService.validateDataSet({ dataSetId }),
                    ]);
                    if (!metadataMatches(dataSetMetadata, requestedMetadata)) {
                        return null;
                    }
                    return {
                        dataSetId,
                        dataSetMetadata,
                        activePieceCount,
                    };
                }
                catch (error) {
                    console.warn(`Skipping data set ${dataSetId} for provider ${providerId}:`, error instanceof Error ? error.message : String(error));
                    return null;
                }
            }));
            for (const result of batchResults) {
                if (result == null)
                    continue;
                if (result.activePieceCount > 0) {
                    selectedDataSet = result;
                    break;
                }
                if (selectedDataSet == null) {
                    selectedDataSet = result;
                }
            }
            if (selectedDataSet != null && selectedDataSet.activePieceCount > 0) {
                break;
            }
        }
        if (selectedDataSet != null) {
            return {
                provider,
                dataSetId: selectedDataSet.dataSetId,
                isExisting: true,
                dataSetMetadata: selectedDataSet.dataSetMetadata,
            };
        }
        return {
            provider,
            dataSetId: -1n,
            isExisting: false,
            dataSetMetadata: requestedMetadata,
        };
    }
    static async resolveByProviderAddress(providerAddress, warmStorageService, spRegistry, clientAddress, requestedMetadata, forceCreateDataSet) {
        const provider = await spRegistry.getProviderByAddress({ address: providerAddress });
        if (provider == null) {
            throw createError('StorageContext', 'resolveByProviderAddress', `Provider ${providerAddress} not found in registry`);
        }
        return await StorageContext.resolveByProviderId(clientAddress, provider.id, requestedMetadata, warmStorageService, spRegistry, forceCreateDataSet);
    }
    static async smartSelectProvider(clientAddress, requestedMetadata, warmStorageService, spRegistry, excludeProviderIds, endorsedProviderIds, forceCreateDataSet) {
        const dataSets = await warmStorageService.getClientDataSetsWithDetails({ address: clientAddress });
        const skipProviderIds = new Set(excludeProviderIds);
        const managedDataSets = dataSets.filter((ps) => ps.isLive &&
            ps.isManaged &&
            ps.pdpEndEpoch === 0n &&
            metadataMatches(ps.metadata, requestedMetadata) &&
            !skipProviderIds.has(ps.providerId));
        if (managedDataSets.length > 0 && !forceCreateDataSet) {
            const sorted = managedDataSets.sort((a, b) => {
                if (a.activePieceCount > 0n && b.activePieceCount === 0n)
                    return -1;
                if (b.activePieceCount > 0n && a.activePieceCount === 0n)
                    return 1;
                return Number(a.pdpVerifierDataSetId - b.pdpVerifierDataSetId);
            });
            async function* generateProviders() {
                for (const dataSet of sorted) {
                    if (skipProviderIds.has(dataSet.providerId)) {
                        continue;
                    }
                    skipProviderIds.add(dataSet.providerId);
                    const provider = await spRegistry.getProvider({ providerId: dataSet.providerId });
                    if (provider == null) {
                        console.warn(`Provider ID ${dataSet.providerId} for data set ${dataSet.pdpVerifierDataSetId} is not currently approved`);
                        continue;
                    }
                    yield provider;
                }
            }
            const selectedProvider = await StorageContext.selectProviderWithPing(generateProviders());
            if (selectedProvider != null) {
                const matchingDataSet = sorted.find((ps) => ps.providerId === selectedProvider.id);
                if (matchingDataSet == null) {
                    console.warn(`Could not match selected provider ${selectedProvider.serviceProvider} (ID: ${selectedProvider.id}) ` +
                        `to existing data sets. Falling back to selecting from all providers.`);
                }
                else {
                    const dataSetMetadata = await warmStorageService.getDataSetMetadata({
                        dataSetId: matchingDataSet.pdpVerifierDataSetId,
                    });
                    return {
                        provider: selectedProvider,
                        dataSetId: matchingDataSet.pdpVerifierDataSetId,
                        isExisting: true,
                        dataSetMetadata,
                    };
                }
            }
        }
        const approvedIds = await warmStorageService.getApprovedProviderIds();
        const approvedProviders = await spRegistry.getProviders({ providerIds: approvedIds });
        const allProviders = approvedProviders.filter((provider) => !excludeProviderIds.includes(provider.id));
        if (allProviders.length === 0) {
            throw createError('StorageContext', 'smartSelectProvider', NO_REMAINING_PROVIDERS_ERROR_MESSAGE);
        }
        let provider;
        if (endorsedProviderIds.size > 0) {
            const [otherProviders, endorsedProviders] = allProviders.reduce((results, provider) => {
                results[endorsedProviderIds.has(provider.id) ? 1 : 0].push(provider);
                return results;
            }, [[], []]);
            provider =
                (await StorageContext.selectRandomProvider(endorsedProviders)) ||
                    (await StorageContext.selectRandomProvider(otherProviders));
        }
        else {
            provider = await StorageContext.selectRandomProvider(allProviders);
        }
        if (provider == null) {
            throw createError('StorageContext', 'selectProviderWithPing', `All ${allProviders.length} providers failed health check. Storage may be temporarily unavailable.`);
        }
        return {
            provider,
            dataSetId: -1n,
            isExisting: false,
            dataSetMetadata: requestedMetadata,
        };
    }
    static async selectRandomProvider(providers) {
        if (providers.length === 0) {
            return null;
        }
        async function* generateRandomProviders() {
            const remaining = [...providers];
            while (remaining.length > 0) {
                const selected = remaining.splice(randIndex(remaining.length), 1)[0];
                yield selected;
            }
        }
        return await StorageContext.selectProviderWithPing(generateRandomProviders());
    }
    static async selectProviderWithPing(providers) {
        for await (const provider of providers) {
            try {
                await SP.ping(provider.pdp.serviceURL);
                return provider;
            }
            catch (error) {
                console.warn(`Provider ${provider.serviceProvider} failed ping test:`, error instanceof Error ? error.message : String(error));
            }
        }
        return null;
    }
    static async performPreflightCheck(options) {
        const { size, withCDN, warmStorageService } = options;
        StorageContext.validateRawSize({ sizeBytes: options.size, context: 'preflightUpload' });
        const allowanceCheck = await warmStorageService.checkAllowanceForStorage({ sizeInBytes: BigInt(size), withCDN });
        return {
            estimatedCost: {
                perEpoch: allowanceCheck.costs.perEpoch,
                perDay: allowanceCheck.costs.perDay,
                perMonth: allowanceCheck.costs.perMonth,
            },
            allowanceCheck: {
                sufficient: allowanceCheck.sufficient,
                message: allowanceCheck.message,
            },
            selectedProvider: null,
            selectedDataSetId: null,
        };
    }
    async preflightUpload(options) {
        const preflightResult = await StorageContext.performPreflightCheck({
            size: options.size,
            withCDN: this._withCDN,
            warmStorageService: this._warmStorageService,
        });
        return preflightResult;
    }
    async upload(data, options) {
        performance.mark('synapse:upload-start');
        const pieceCid = options?.pieceCid;
        const uploadId = Symbol('upload');
        this._activeUploads.add(uploadId);
        try {
            let uploadResult;
            try {
                uploadResult = await SP.uploadPieceStreaming({
                    serviceURL: this._pdpEndpoint,
                    data,
                    ...options,
                    pieceCid,
                });
            }
            catch (error) {
                throw createError('StorageContext', 'uploadPiece', 'Failed to upload piece to service provider', error);
            }
            await SP.findPiece({
                serviceURL: this._pdpEndpoint,
                pieceCid: uploadResult.pieceCid,
                retry: true,
            });
            this._activeUploads.delete(uploadId);
            if (options?.onUploadComplete != null) {
                options.onUploadComplete(uploadResult.pieceCid);
            }
            if (options?.metadata != null) {
                pieceMetadataObjectToEntry(options.metadata);
            }
            const finalPieceId = await new Promise((resolve, reject) => {
                this._pendingPieces.push({
                    pieceCid: uploadResult.pieceCid,
                    resolve,
                    reject,
                    callbacks: options,
                    metadata: options?.metadata,
                });
                setTimeout(() => {
                    void this._processPendingPieces().catch((error) => {
                        console.error('Failed to process pending pieces batch:', error);
                    });
                }, 0);
            });
            performance.mark('synapse:upload-end');
            performance.measure('synapse:upload', 'synapse:upload-start', 'synapse:upload-end');
            return {
                pieceCid: uploadResult.pieceCid,
                size: uploadResult.size,
                pieceId: finalPieceId,
            };
        }
        finally {
            this._activeUploads.delete(uploadId);
        }
    }
    async _processPendingPieces() {
        if (this._isProcessing || this._pendingPieces.length === 0) {
            return;
        }
        this._isProcessing = true;
        const uploadsToWaitFor = new Set(this._activeUploads);
        if (uploadsToWaitFor.size > 0 && this._pendingPieces.length < this._uploadBatchSize) {
            const waitStart = Date.now();
            const pollInterval = 200;
            while (uploadsToWaitFor.size > 0 && Date.now() - waitStart < this._uploadBatchWaitTimeout) {
                for (const uploadId of uploadsToWaitFor) {
                    if (!this._activeUploads.has(uploadId)) {
                        uploadsToWaitFor.delete(uploadId);
                    }
                }
                if (uploadsToWaitFor.size > 0) {
                    await new Promise((resolve) => setTimeout(resolve, pollInterval));
                }
            }
            const waited = Date.now() - waitStart;
            if (waited > pollInterval) {
                console.debug(`Waited ${waited}ms for ${uploadsToWaitFor.size} active upload(s) to complete`);
            }
        }
        const batch = this._pendingPieces.splice(0, this._uploadBatchSize);
        try {
            const pieceCids = batch.map((item) => item.pieceCid);
            const confirmedPieceIds = [];
            const addedPieceRecords = pieceCids.map((pieceCid) => ({ pieceCid }));
            if (this.dataSetId) {
                const [, clientDataSetId] = await Promise.all([
                    this._warmStorageService.validateDataSet({ dataSetId: this.dataSetId }),
                    this.getClientDataSetId(),
                ]);
                const addPiecesResult = await SP.addPieces(this._client, {
                    dataSetId: this.dataSetId,
                    clientDataSetId,
                    pieces: batch.map((item) => ({ pieceCid: item.pieceCid, metadata: item.metadata })),
                    serviceURL: this._pdpEndpoint,
                });
                batch.forEach((item) => {
                    item.callbacks?.onPiecesAdded?.(addPiecesResult.txHash, addedPieceRecords);
                });
                const confirmation = await SP.waitForAddPieces(addPiecesResult);
                confirmedPieceIds.push(...confirmation.confirmedPieceIds);
                const confirmedPieceRecords = confirmedPieceIds.map((pieceId, index) => ({
                    pieceId,
                    pieceCid: pieceCids[index],
                }));
                batch.forEach((item) => {
                    item.callbacks?.onPiecesConfirmed?.(this.dataSetId, confirmedPieceRecords);
                });
            }
            else {
                const result = await SP.createDataSetAndAddPieces(this._client, {
                    cdn: this._withCDN,
                    payee: this._provider.serviceProvider,
                    payer: this._client.account.address,
                    recordKeeper: this._chain.contracts.fwss.address,
                    pieces: batch.map((item) => ({ pieceCid: item.pieceCid, metadata: item.metadata })),
                    metadata: this._dataSetMetadata,
                    serviceURL: this._pdpEndpoint,
                });
                batch.forEach((item) => {
                    item.callbacks?.onPiecesAdded?.(result.txHash, addedPieceRecords);
                });
                const confirmation = await SP.waitForCreateDataSetAddPieces(result);
                this._dataSetId = confirmation.dataSetId;
                confirmedPieceIds.push(...confirmation.piecesIds);
                const confirmedPieceRecords = confirmedPieceIds.map((pieceId, index) => ({
                    pieceId,
                    pieceCid: pieceCids[index],
                }));
                batch.forEach((item) => {
                    item.callbacks?.onPiecesConfirmed?.(this.dataSetId, confirmedPieceRecords);
                });
            }
            batch.forEach((item, index) => {
                const pieceId = confirmedPieceIds[index];
                if (pieceId == null) {
                    throw createError('StorageContext', 'addPieces', `Server did not return piece ID for piece at index ${index}`);
                }
                item.resolve(pieceId);
            });
        }
        catch (error) {
            const finalError = createError('StorageContext', 'addPieces', 'Failed to add piece to data set', error);
            batch.forEach((item) => {
                item.reject(finalError);
            });
        }
        finally {
            this._isProcessing = false;
            if (this._pendingPieces.length > 0) {
                void this._processPendingPieces().catch((error) => {
                    console.error('Failed to process pending pieces batch:', error);
                });
            }
        }
    }
    async download(options) {
        return this._synapse.storage.download({
            pieceCid: options.pieceCid,
            providerAddress: this._provider.serviceProvider,
            withCDN: options?.withCDN ?? this._withCDN,
        });
    }
    async getProviderInfo() {
        return await this._synapse.getProviderInfo(this.serviceProvider);
    }
    async getScheduledRemovals() {
        if (this._dataSetId == null) {
            return [];
        }
        return await PDPVerifier.getScheduledRemovals(this._client, { dataSetId: this._dataSetId });
    }
    async *getPieces(options = {}) {
        if (this._dataSetId == null) {
            return;
        }
        const batchSize = options?.batchSize ?? 100n;
        let offset = 0n;
        let hasMore = true;
        while (hasMore) {
            const result = await PDPVerifier.getActivePieces(this._client, {
                dataSetId: this._dataSetId,
                offset,
                limit: batchSize,
            });
            for (let i = 0; i < result.pieces.length; i++) {
                yield {
                    pieceCid: result.pieces[i].cid,
                    pieceId: result.pieces[i].id,
                };
            }
            hasMore = result.hasMore;
            offset += batchSize;
        }
    }
    async _getPieceIdByCID(pieceCid) {
        if (this.dataSetId == null) {
            throw createError('StorageContext', 'getPieceIdByCID', 'Data set not found');
        }
        const parsedPieceCID = asPieceCID(pieceCid);
        if (parsedPieceCID == null) {
            throw createError('StorageContext', 'deletePiece', 'Invalid PieceCID provided');
        }
        const dataSetData = await SP.getDataSet({
            serviceURL: this._pdpEndpoint,
            dataSetId: this.dataSetId,
        });
        const pieceData = dataSetData.pieces.find((piece) => piece.pieceCid.toString() === parsedPieceCID.toString());
        if (pieceData == null) {
            throw createError('StorageContext', 'deletePiece', 'Piece not found in data set');
        }
        return pieceData.pieceId;
    }
    async deletePiece(options) {
        const { piece } = options;
        if (this.dataSetId == null) {
            throw createError('StorageContext', 'deletePiece', 'Data set not found');
        }
        const pieceId = typeof piece === 'bigint' ? piece : await this._getPieceIdByCID(piece);
        const clientDataSetId = await this.getClientDataSetId();
        const { hash } = await schedulePieceDeletion(this._synapse.client, {
            serviceURL: this._pdpEndpoint,
            dataSetId: this.dataSetId,
            pieceId: pieceId,
            clientDataSetId: clientDataSetId,
        });
        return hash;
    }
    async hasPiece(options) {
        const { pieceCid } = options;
        const parsedPieceCID = asPieceCID(pieceCid);
        if (parsedPieceCID == null) {
            return false;
        }
        try {
            await SP.findPiece({
                serviceURL: this._pdpEndpoint,
                pieceCid: parsedPieceCID,
            });
            return true;
        }
        catch {
            return false;
        }
    }
    async pieceStatus(options) {
        if (this.dataSetId == null) {
            throw createError('StorageContext', 'pieceStatus', 'Data set not found');
        }
        const parsedPieceCID = asPieceCID(options.pieceCid);
        if (parsedPieceCID == null) {
            throw createError('StorageContext', 'pieceStatus', 'Invalid PieceCID provided');
        }
        const [exists, dataSetData, currentEpoch] = await Promise.all([
            this.hasPiece({ pieceCid: parsedPieceCID }),
            SP.getDataSet({
                serviceURL: this._pdpEndpoint,
                dataSetId: this.dataSetId,
            }),
            getBlockNumber(this._client),
        ]);
        let retrievalUrl = null;
        let pieceId;
        let lastProven = null;
        let nextProofDue = null;
        let inChallengeWindow = false;
        let hoursUntilChallengeWindow = 0;
        let isProofOverdue = false;
        if (exists) {
            const [providerInfo, pdpConfig] = await Promise.all([
                this.getProviderInfo().catch(() => null),
                dataSetData != null
                    ? this._warmStorageService.getPDPConfig().catch((error) => {
                        console.debug('Failed to get PDP config:', error);
                        return null;
                    })
                    : Promise.resolve(null),
            ]);
            if (providerInfo != null) {
                retrievalUrl = createPieceUrlPDP({
                    cid: parsedPieceCID.toString(),
                    serviceURL: providerInfo.pdp.serviceURL,
                });
            }
            if (dataSetData != null && pdpConfig != null) {
                const pieceData = dataSetData.pieces.find((piece) => piece.pieceCid.toString() === parsedPieceCID.toString());
                if (pieceData != null) {
                    pieceId = pieceData.pieceId;
                    if (dataSetData.nextChallengeEpoch > 0) {
                        const challengeWindowStart = dataSetData.nextChallengeEpoch;
                        const provingDeadline = challengeWindowStart + Number(pdpConfig.challengeWindowSize);
                        nextProofDue = epochToDate(provingDeadline, this._chain.genesisTimestamp);
                        const lastProvenDate = calculateLastProofDate(dataSetData.nextChallengeEpoch, Number(pdpConfig.maxProvingPeriod), this._chain.genesisTimestamp);
                        if (lastProvenDate != null) {
                            lastProven = lastProvenDate;
                        }
                        inChallengeWindow = Number(currentEpoch) >= challengeWindowStart && Number(currentEpoch) < provingDeadline;
                        isProofOverdue = Number(currentEpoch) >= provingDeadline;
                        if (Number(currentEpoch) < challengeWindowStart) {
                            const timeUntil = timeUntilEpoch(challengeWindowStart, Number(currentEpoch));
                            hoursUntilChallengeWindow = timeUntil.hours;
                        }
                    }
                    else {
                        console.debug('Data set has nextChallengeEpoch=0, may have just been proven');
                    }
                }
            }
        }
        return {
            exists,
            dataSetLastProven: lastProven,
            dataSetNextProofDue: nextProofDue,
            retrievalUrl,
            pieceId,
            inChallengeWindow,
            hoursUntilChallengeWindow,
            isProofOverdue,
        };
    }
    async terminate() {
        if (this.dataSetId == null) {
            throw createError('StorageContext', 'terminate', 'Data set not found');
        }
        return this._synapse.storage.terminateDataSet({ dataSetId: this.dataSetId });
    }
}
