// @filoz/synapse-sdk/dist/src/storage/manager.js
import { asPieceCID, downloadAndValidate } from '@filoz/synapse-core/piece';
import { randIndex } from '@filoz/synapse-core/utils';
import { zeroAddress } from 'viem';
import { SPRegistryService } from "../sp-registry/index.js";
import { combineMetadata, createError, METADATA_KEYS, metadataMatches, SIZE_CONSTANTS, TIME_CONSTANTS, } from "../utils/index.js";
import { StorageContext } from "./context.js";
export class StorageManager {
    _synapse;
    _warmStorageService;
    _pieceRetriever;
    _withCDN;
    _defaultContexts;
    constructor(options) {
        this._synapse = options.synapse;
        this._warmStorageService = options.warmStorageService;
        this._pieceRetriever = options.pieceRetriever;
        this._withCDN = options.withCDN;
    }
    async upload(data, options) {
        if (options?.context != null || options?.contexts != null) {
            const invalidOptions = [];
            if (options.providerId !== undefined)
                invalidOptions.push('providerId');
            if (options.providerAddress !== undefined)
                invalidOptions.push('providerAddress');
            if (options.dataSetId !== undefined)
                invalidOptions.push('dataSetId');
            if (options.withCDN !== undefined)
                invalidOptions.push('withCDN');
            if (options.forceCreateDataSet !== undefined)
                invalidOptions.push('forceCreateDataSet');
            if (options.uploadBatchSize !== undefined)
                invalidOptions.push('uploadBatchSize');
            if (invalidOptions.length > 0) {
                throw createError('StorageManager', 'upload', `Cannot specify both 'context' and other options: ${invalidOptions.join(', ')}`);
            }
        }
        if (options?.contexts != null && options.contexts.length > 0) {
            if (options?.context != null) {
                throw createError('StorageManager', 'upload', "Cannot specify both 'context' and 'contexts'");
            }
        }
        const contexts = options?.contexts ??
            (options?.context
                ? [options.context]
                : await this.createContexts({
                    withCDN: options?.withCDN,
                    count: 1,
                    uploadBatchSize: options?.uploadBatchSize,
                    forceCreateDataSets: options?.forceCreateDataSet,
                    metadata: options?.metadata,
                    excludeProviderIds: options?.excludeProviderIds,
                    providerIds: options?.providerId ? [options.providerId] : undefined,
                    dataSetIds: options?.dataSetId ? [options.dataSetId] : undefined,
                    callbacks: options?.callbacks,
                }));
        if (contexts.length > 1) {
            if (data instanceof ReadableStream) {
                throw createError('StorageManager', 'upload', 'Streaming uploads are not supported for multiple contexts');
            }
            return Promise.all(contexts.map((context) => context.upload(data, {
                ...options?.callbacks,
                metadata: options?.metadata,
                signal: options?.signal,
            }))).then((results) => results[0]);
        }
        else {
            const context = contexts[0];
            return context.upload(data, {
                ...options?.callbacks,
                metadata: options?.metadata,
                signal: options?.signal,
            });
        }
    }
    async download(options) {
        if (options?.context != null) {
            const invalidOptions = [];
            if (options.providerAddress !== undefined)
                invalidOptions.push('providerAddress');
            if (options.withCDN !== undefined)
                invalidOptions.push('withCDN');
            if (invalidOptions.length > 0) {
                throw createError('StorageManager', 'download', `Cannot specify both 'context' and other options: ${invalidOptions.join(', ')}`);
            }
            return await options.context.download({
                pieceCid: options.pieceCid,
                withCDN: options.withCDN ?? this._withCDN,
            });
        }
        const parsedPieceCID = asPieceCID(options.pieceCid);
        if (parsedPieceCID == null) {
            throw createError('StorageManager', 'download', `Invalid PieceCID: ${String(options.pieceCid)}`);
        }
        const withCDN = options?.withCDN ?? this._withCDN;
        let finalProviderAddress = options?.providerAddress;
        if (this._defaultContexts != null && !withCDN && finalProviderAddress == null) {
            const contextsWithoutCDN = this._defaultContexts.filter((context) => context.withCDN === false);
            const contextsHavePiece = await Promise.all(contextsWithoutCDN.map((context) => context.hasPiece({ pieceCid: parsedPieceCID })));
            const defaultContextsWithPiece = contextsWithoutCDN.filter((_context, i) => contextsHavePiece[i]);
            if (defaultContextsWithPiece.length > 0) {
                finalProviderAddress =
                    defaultContextsWithPiece[randIndex(defaultContextsWithPiece.length)].provider.serviceProvider;
            }
        }
        const clientAddress = this._synapse.client.account.address;
        const response = await this._pieceRetriever.fetchPiece({
            pieceCid: parsedPieceCID,
            client: clientAddress,
            providerAddress: finalProviderAddress,
            withCDN,
        });
        return await downloadAndValidate(response, parsedPieceCID);
    }
    async preflightUpload(options) {
        let withCDN = options?.withCDN ?? this._withCDN;
        if (options?.metadata != null && METADATA_KEYS.WITH_CDN in options.metadata) {
            const value = options.metadata[METADATA_KEYS.WITH_CDN];
            if (value !== '') {
                console.warn(`Warning: withCDN metadata entry has unexpected value "${value}". Expected empty string.`);
            }
            withCDN = true;
        }
        return await StorageContext.performPreflightCheck({
            warmStorageService: this._warmStorageService,
            size: options.size,
            withCDN,
        });
    }
    async createContexts(options) {
        const withCDN = options?.withCDN ?? this._withCDN;
        const canUseDefault = options == null ||
            (options.providerIds == null &&
                options.dataSetIds == null &&
                options.forceCreateDataSets !== true &&
                options.uploadBatchSize == null);
        if (this._defaultContexts != null) {
            const expectedSize = options?.count ?? 2;
            if (this._defaultContexts.length === expectedSize &&
                this._defaultContexts.every((context) => options?.excludeProviderIds?.includes(context.provider.id) !== true)) {
                const requestedMetadata = combineMetadata(options?.metadata, withCDN);
                if (this._defaultContexts.every((defaultContext) => metadataMatches(defaultContext.dataSetMetadata, requestedMetadata))) {
                    if (options?.callbacks != null) {
                        for (const defaultContext of this._defaultContexts) {
                            try {
                                options.callbacks.onProviderSelected?.(defaultContext.provider);
                            }
                            catch (error) {
                                console.error('Error in onProviderSelected callback:', error);
                            }
                            if (defaultContext.dataSetId != null) {
                                try {
                                    options.callbacks.onDataSetResolved?.({
                                        isExisting: true,
                                        dataSetId: defaultContext.dataSetId,
                                        provider: defaultContext.provider,
                                    });
                                }
                                catch (error) {
                                    console.error('Error in onDataSetResolved callback:', error);
                                }
                            }
                        }
                    }
                    return this._defaultContexts;
                }
            }
        }
        const contexts = await StorageContext.createContexts({
            synapse: this._synapse,
            warmStorageService: this._warmStorageService,
            ...options,
            withCDN,
        });
        if (canUseDefault) {
            this._defaultContexts = contexts;
        }
        return contexts;
    }
    async createContext(options) {
        const effectiveWithCDN = options?.withCDN ?? this._withCDN;
        const canUseDefault = options == null ||
            (options.providerId == null &&
                options.providerAddress == null &&
                options.dataSetId == null &&
                options.forceCreateDataSet !== true &&
                options.uploadBatchSize == null);
        if (canUseDefault && this._defaultContexts != null) {
            const requestedMetadata = combineMetadata(options?.metadata, effectiveWithCDN);
            for (const defaultContext of this._defaultContexts) {
                if (options?.excludeProviderIds?.includes(defaultContext.provider.id)) {
                    continue;
                }
                if (!metadataMatches(defaultContext.dataSetMetadata, requestedMetadata)) {
                    continue;
                }
                if (options?.callbacks != null) {
                    try {
                        options.callbacks.onProviderSelected?.(defaultContext.provider);
                    }
                    catch (error) {
                        console.error('Error in onProviderSelected callback:', error);
                    }
                    if (defaultContext.dataSetId != null) {
                        try {
                            options.callbacks.onDataSetResolved?.({
                                isExisting: true,
                                dataSetId: defaultContext.dataSetId,
                                provider: defaultContext.provider,
                            });
                        }
                        catch (error) {
                            console.error('Error in onDataSetResolved callback:', error);
                        }
                    }
                }
                return defaultContext;
            }
        }
        const context = await StorageContext.create({
            synapse: this._synapse,
            warmStorageService: this._warmStorageService,
            ...options,
            withCDN: effectiveWithCDN,
        });
        if (canUseDefault) {
            this._defaultContexts = [context];
        }
        return context;
    }
    async getDefaultContext() {
        return await this.createContext();
    }
    async findDataSets(options = {}) {
        const { address = this._synapse.client.account.address } = options;
        return await this._warmStorageService.getClientDataSetsWithDetails({ address });
    }
    async terminateDataSet(options) {
        return this._warmStorageService.terminateDataSet(options);
    }
    async getStorageInfo() {
        const chain = this._synapse.client.chain;
        try {
            const getOptionalAllowances = async () => {
                try {
                    const approval = await this._synapse.payments.serviceApproval();
                    return {
                        service: chain.contracts.fwss.address,
                        isApproved: approval.isApproved,
                        rateAllowance: approval.rateAllowance,
                        lockupAllowance: approval.lockupAllowance,
                        rateUsed: approval.rateUsage,
                        lockupUsed: approval.lockupUsage,
                    };
                }
                catch {
                    return null;
                }
            };
            const spRegistry = new SPRegistryService({ client: this._synapse.client });
            const [pricingData, approvedIds, allowances] = await Promise.all([
                this._warmStorageService.getServicePrice(),
                this._warmStorageService.getApprovedProviderIds(),
                getOptionalAllowances(),
            ]);
            const providers = await spRegistry.getProviders({ providerIds: approvedIds });
            const epochsPerMonth = BigInt(pricingData.epochsPerMonth);
            const noCDNPerEpoch = BigInt(pricingData.pricePerTiBPerMonthNoCDN) / epochsPerMonth;
            const withCDNPerEpoch = BigInt(pricingData.pricePerTiBPerMonthNoCDN) / epochsPerMonth;
            const noCDNPerDay = BigInt(pricingData.pricePerTiBPerMonthNoCDN) / TIME_CONSTANTS.DAYS_PER_MONTH;
            const withCDNPerDay = BigInt(pricingData.pricePerTiBPerMonthNoCDN) / TIME_CONSTANTS.DAYS_PER_MONTH;
            const validProviders = providers.filter((p) => p.serviceProvider !== zeroAddress);
            return {
                pricing: {
                    noCDN: {
                        perTiBPerMonth: BigInt(pricingData.pricePerTiBPerMonthNoCDN),
                        perTiBPerDay: noCDNPerDay,
                        perTiBPerEpoch: noCDNPerEpoch,
                    },
                    withCDN: {
                        perTiBPerMonth: BigInt(pricingData.pricePerTiBPerMonthNoCDN),
                        perTiBPerDay: withCDNPerDay,
                        perTiBPerEpoch: withCDNPerEpoch,
                    },
                    tokenAddress: pricingData.tokenAddress,
                    tokenSymbol: 'USDFC',
                },
                providers: validProviders,
                serviceParameters: {
                    epochsPerMonth,
                    epochsPerDay: TIME_CONSTANTS.EPOCHS_PER_DAY,
                    epochDuration: TIME_CONSTANTS.EPOCH_DURATION,
                    minUploadSize: SIZE_CONSTANTS.MIN_UPLOAD_SIZE,
                    maxUploadSize: SIZE_CONSTANTS.MAX_UPLOAD_SIZE,
                },
                allowances,
            };
        }
        catch (error) {
            throw new Error(`Failed to get storage service information: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
}
