// @filoz/synapse-sdk/dist/src/warm-storage/service.js
import { asChain } from '@filoz/synapse-core/chains';
import * as Pay from '@filoz/synapse-core/pay';
import * as PDPVerifier from '@filoz/synapse-core/pdp-verifier';
import { dataSetLiveCall, getDataSetListenerCall } from '@filoz/synapse-core/pdp-verifier';
import { metadataArrayToObject } from '@filoz/synapse-core/utils';
import { addApprovedProvider, getAllDataSetMetadata, getAllDataSetMetadataCall, getAllPieceMetadata, getApprovedProviders, getClientDataSets, getDataSet, getServicePrice, removeApprovedProvider, terminateService, } from '@filoz/synapse-core/warm-storage';
import { createClient, http, isAddressEqual, } from 'viem';
import { multicall, readContract, simulateContract, writeContract } from 'viem/actions';
import { DEFAULT_CHAIN, METADATA_KEYS, SIZE_CONSTANTS, TIME_CONSTANTS } from "../utils/constants.js";
export class WarmStorageService {
    _client;
    _chain;
    constructor(options) {
        this._client = options.client;
        this._chain = asChain(options.client.chain);
    }
    static create(options) {
        const client = createClient({
            chain: options.chain ?? DEFAULT_CHAIN,
            transport: options.transport ?? http(),
            account: options.account,
            name: 'WarmStorageService',
            key: 'warm-storage-service',
        });
        if (client.account.type === 'json-rpc' && client.transport.type !== 'custom') {
            throw new Error('Transport must be a custom transport. See https://viem.sh/docs/clients/transports/custom.');
        }
        return new WarmStorageService({ client });
    }
    async getDataSet(options) {
        return getDataSet(this._client, options);
    }
    async getClientDataSets(options) {
        return getClientDataSets(this._client, options);
    }
    async getClientDataSetsWithDetails(options) {
        const { address = this._client.account.address, onlyManaged = false } = options;
        const ids = await readContract(this._client, {
            address: this._chain.contracts.fwssView.address,
            abi: this._chain.contracts.fwssView.abi,
            functionName: 'clientDataSets',
            args: [address],
        });
        if (ids.length === 0)
            return [];
        const enhancedDataSetsPromises = ids.map(async (dataSetId) => {
            try {
                const base = await this.getDataSet({ dataSetId });
                if (base == null)
                    return null;
                const [isLive, listener, metadata] = await multicall(this._client, {
                    allowFailure: false,
                    contracts: [
                        dataSetLiveCall({
                            chain: this._client.chain,
                            dataSetId: dataSetId,
                        }),
                        getDataSetListenerCall({
                            chain: this._client.chain,
                            dataSetId: dataSetId,
                        }),
                        getAllDataSetMetadataCall({
                            chain: this._client.chain,
                            dataSetId: dataSetId,
                        }),
                    ],
                });
                const isManaged = isAddressEqual(listener, this._chain.contracts.fwss.address);
                if (onlyManaged && !isManaged) {
                    return null;
                }
                const activePieceCount = isLive ? await PDPVerifier.getActivePieceCount(this._client, { dataSetId }) : 0n;
                return {
                    ...base,
                    pdpVerifierDataSetId: dataSetId,
                    activePieceCount,
                    isLive,
                    isManaged,
                    withCDN: base.cdnRailId > 0 && metadata[0].includes(METADATA_KEYS.WITH_CDN),
                    metadata: metadataArrayToObject(metadata),
                };
            }
            catch (error) {
                throw new Error(`Failed to get details for data set ${dataSetId}: ${error instanceof Error ? error.message : String(error)}`);
            }
        });
        const results = await Promise.all(enhancedDataSetsPromises);
        return results.filter((result) => result !== null);
    }
    async validateDataSet(options) {
        const [isLive, listener] = await multicall(this._client, {
            allowFailure: false,
            contracts: [
                dataSetLiveCall({
                    chain: this._client.chain,
                    dataSetId: options.dataSetId,
                }),
                getDataSetListenerCall({
                    chain: this._client.chain,
                    dataSetId: options.dataSetId,
                }),
            ],
        });
        if (!isLive) {
            throw new Error(`Data set ${options.dataSetId} does not exist or is not live`);
        }
        if (!isAddressEqual(listener, this._chain.contracts.fwss.address)) {
            throw new Error(`Data set ${options.dataSetId} is not managed by this WarmStorage contract (${this._chain.contracts.fwss.address}), managed by ${String(listener)}`);
        }
    }
    async getActivePieceCount(options) {
        return PDPVerifier.getActivePieceCount(this._client, { dataSetId: options.dataSetId });
    }
    async getDataSetMetadata(options) {
        return getAllDataSetMetadata(this._client, { dataSetId: options.dataSetId });
    }
    async getDataSetMetadataByKey(options) {
        const [exists, value] = await readContract(this._client, {
            address: this._chain.contracts.fwssView.address,
            abi: this._chain.contracts.fwssView.abi,
            functionName: 'getDataSetMetadata',
            args: [options.dataSetId, options.key],
        });
        return exists ? value : null;
    }
    async getPieceMetadata(options) {
        return getAllPieceMetadata(this._client, options);
    }
    async getPieceMetadataByKey(options) {
        const [exists, value] = await readContract(this._client, {
            address: this._chain.contracts.fwssView.address,
            abi: this._chain.contracts.fwssView.abi,
            functionName: 'getPieceMetadata',
            args: [options.dataSetId, options.pieceId, options.key],
        });
        return exists ? value : null;
    }
    async getServicePrice() {
        return getServicePrice(this._client);
    }
    async calculateStorageCost(options) {
        const servicePriceInfo = await this.getServicePrice();
        const pricePerEpoch = (servicePriceInfo.pricePerTiBPerMonthNoCDN * options.sizeInBytes) /
            (SIZE_CONSTANTS.TiB * servicePriceInfo.epochsPerMonth);
        const costs = {
            perEpoch: pricePerEpoch,
            perDay: pricePerEpoch * BigInt(TIME_CONSTANTS.EPOCHS_PER_DAY),
            perMonth: pricePerEpoch * servicePriceInfo.epochsPerMonth,
        };
        return {
            ...costs,
            withCDN: costs,
        };
    }
    async checkAllowanceForStorage(options) {
        const [approval, costs] = await Promise.all([
            Pay.operatorApprovals(this._client, {
                address: options.address ?? this._client.account.address,
                operator: options.operator,
            }),
            this.calculateStorageCost({ sizeInBytes: options.sizeInBytes }),
        ]);
        const selectedCosts = options.withCDN ? costs.withCDN : costs;
        const rateNeeded = selectedCosts.perEpoch;
        const lockupPeriod = (options.lockupDays ?? TIME_CONSTANTS.DEFAULT_LOCKUP_DAYS) * TIME_CONSTANTS.EPOCHS_PER_DAY;
        const lockupNeeded = rateNeeded * lockupPeriod;
        const totalRateNeeded = approval.rateUsage + rateNeeded;
        const totalLockupNeeded = approval.lockupUsage + lockupNeeded;
        const sufficient = approval.rateAllowance >= totalRateNeeded && approval.lockupAllowance >= totalLockupNeeded;
        const rateAllowanceNeeded = totalRateNeeded > approval.rateAllowance ? totalRateNeeded - approval.rateAllowance : 0n;
        const lockupAllowanceNeeded = totalLockupNeeded > approval.lockupAllowance ? totalLockupNeeded - approval.lockupAllowance : 0n;
        let message;
        if (!sufficient) {
            const needsRate = rateAllowanceNeeded > 0n;
            const needsLockup = lockupAllowanceNeeded > 0n;
            if (needsRate && needsLockup) {
                message = 'Insufficient rate and lockup allowances';
            }
            else if (needsRate) {
                message = 'Insufficient rate allowance';
            }
            else if (needsLockup) {
                message = 'Insufficient lockup allowance';
            }
        }
        return {
            rateAllowanceNeeded,
            lockupAllowanceNeeded,
            currentRateAllowance: approval.rateAllowance,
            currentLockupAllowance: approval.lockupAllowance,
            currentRateUsed: approval.rateUsage,
            currentLockupUsed: approval.lockupUsage,
            sufficient,
            message,
            costs: selectedCosts,
            depositAmountNeeded: lockupNeeded,
        };
    }
    async prepareStorageUpload(options) {
        const address = options.address ?? this._client.account.address;
        const [costs, allowanceCheck] = await Promise.all([
            this.calculateStorageCost({ sizeInBytes: options.dataSize }),
            this.checkAllowanceForStorage({
                sizeInBytes: options.dataSize,
                withCDN: options.withCDN ?? false,
                address: options.address,
                operator: options.operator,
                lockupDays: options.lockupDays,
            }),
        ]);
        const selectedCosts = (options.withCDN ?? false) ? costs.withCDN : costs;
        const actions = [];
        const accountInfo = await Pay.accounts(this._client, { address });
        const requiredBalance = selectedCosts.perMonth;
        if (accountInfo.availableFunds < requiredBalance) {
            const depositAmount = requiredBalance - accountInfo.availableFunds;
            actions.push({
                type: 'deposit',
                description: `Deposit ${depositAmount} USDFC to payments contract`,
                execute: async () => await Pay.deposit(this._client, { amount: depositAmount }),
            });
        }
        if (!allowanceCheck.sufficient) {
            actions.push({
                type: 'approveService',
                description: `Approve service with rate allowance ${allowanceCheck.rateAllowanceNeeded} and lockup allowance ${allowanceCheck.lockupAllowanceNeeded}`,
                execute: async () => await Pay.setOperatorApproval(this._client, {
                    approve: true,
                }),
            });
        }
        return {
            estimatedCost: {
                perEpoch: selectedCosts.perEpoch,
                perDay: selectedCosts.perDay,
                perMonth: selectedCosts.perMonth,
            },
            allowanceCheck: {
                sufficient: allowanceCheck.sufficient,
                message: allowanceCheck.sufficient
                    ? undefined
                    : `Insufficient allowances: rate needed ${allowanceCheck.rateAllowanceNeeded}, lockup needed ${allowanceCheck.lockupAllowanceNeeded}`,
            },
            actions,
        };
    }
    async terminateDataSet(options) {
        return terminateService(this._client, { dataSetId: options.dataSetId });
    }
    async addApprovedProvider(options) {
        return addApprovedProvider(this._client, { providerId: options.providerId });
    }
    async removeApprovedProvider(options) {
        const approvedIds = await getApprovedProviders(this._client);
        const index = approvedIds.indexOf(options.providerId);
        if (index === -1) {
            throw new Error(`Provider ${options.providerId} is not in the approved list`);
        }
        return removeApprovedProvider(this._client, { providerId: options.providerId, index: BigInt(index) });
    }
    async getApprovedProviderIds() {
        return getApprovedProviders(this._client);
    }
    async isProviderIdApproved(options) {
        return readContract(this._client, {
            address: this._chain.contracts.fwssView.address,
            abi: this._chain.contracts.fwssView.abi,
            functionName: 'isProviderApproved',
            args: [options.providerId],
        });
    }
    async getOwner() {
        return readContract(this._client, {
            address: this._chain.contracts.fwss.address,
            abi: this._chain.contracts.fwss.abi,
            functionName: 'owner',
        });
    }
    async isOwner(options) {
        const ownerAddress = await this.getOwner();
        return isAddressEqual(options.address, ownerAddress);
    }
    async getPDPConfig() {
        const [maxProvingPeriod, challengeWindowSize, challengesPerProof, initChallengeWindowStart] = await readContract(this._client, {
            address: this._chain.contracts.fwssView.address,
            abi: this._chain.contracts.fwssView.abi,
            functionName: 'getPDPConfig',
        });
        return {
            maxProvingPeriod: maxProvingPeriod,
            challengeWindowSize: challengeWindowSize,
            challengesPerProof: challengesPerProof,
            initChallengeWindowStart: initChallengeWindowStart,
        };
    }
    async topUpCDNPaymentRails(options) {
        if (options.cdnAmountToAdd < 0n || options.cacheMissAmountToAdd < 0n) {
            throw new Error('Top up amounts must be positive');
        }
        if (options.cdnAmountToAdd === 0n && options.cacheMissAmountToAdd === 0n) {
            throw new Error('At least one top up amount must be >0');
        }
        const { request } = await simulateContract(this._client, {
            address: this._chain.contracts.fwss.address,
            abi: this._chain.contracts.fwss.abi,
            functionName: 'topUpCDNPaymentRails',
            args: [options.dataSetId, options.cdnAmountToAdd, options.cacheMissAmountToAdd],
        });
        const hash = await writeContract(this._client, request);
        return hash;
    }
}
