// @filoz/synapse-core/dist/src/pay/accounts.js
import { getBlockNumber, readContract } from 'viem/actions';
import { asChain } from "../chains.js";
export async function accounts(client, options) {
    const currentEpoch = options.blockNumber ??
        (await getBlockNumber(client, {
            cacheTime: 0,
        }));
    const data = await readContract(client, accountsCall({
        chain: client.chain,
        token: options.token,
        address: options.address,
        contractAddress: options.contractAddress,
    }));
    return parseAccounts(data, currentEpoch);
}
export function accountsCall(options) {
    const chain = asChain(options.chain);
    const token = options.token ?? chain.contracts.usdfc.address;
    return {
        abi: chain.contracts.filecoinPay.abi,
        address: options.contractAddress ?? chain.contracts.filecoinPay.address,
        functionName: 'accounts',
        args: [token, options.address],
    };
}
export function parseAccounts(data, currentEpoch) {
    const [funds, lockupCurrent, lockupRate, lockupLastSettledAt] = data;
    const epochSinceSettlement = currentEpoch - lockupLastSettledAt;
    const actualLockup = lockupCurrent + epochSinceSettlement * lockupRate;
    const availableFunds = funds - actualLockup;
    return {
        funds: data[0],
        lockupCurrent: data[1],
        lockupRate: data[2],
        lockupLastSettledAt: data[3],
        availableFunds: availableFunds < 0n ? 0n : availableFunds,
    };
}
