// @filoz/synapse-core/dist/src/pay/deposit.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
import * as erc20 from "../erc20/index.js";
import { ValidationError } from "../errors/base.js";
import { InsufficientAllowanceError, InsufficientBalanceError } from "../errors/pay.js";
export async function deposit(client, options) {
    const balance = await erc20.balance(client, {
        address: client.account.address,
        token: options.token,
    });
    if (balance.value < options.amount) {
        throw new InsufficientBalanceError(balance.value, options.amount);
    }
    if (balance.allowance < options.amount) {
        throw new InsufficientAllowanceError(balance.allowance, options.amount);
    }
    const { request } = await simulateContract(client, depositCall({
        chain: client.chain,
        token: options.token,
        to: options.to ?? client.account.address,
        amount: options.amount,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function depositSync(client, options) {
    const hash = await deposit(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractDepositEvent(receipt.logs);
    return { receipt, event };
}
export function depositCall(options) {
    const chain = asChain(options.chain);
    const token = options.token ?? chain.contracts.usdfc.address;
    if (options.amount <= 0n) {
        throw new ValidationError('Deposit amount must be greater than 0');
    }
    return {
        abi: chain.contracts.filecoinPay.abi,
        address: options.contractAddress ?? chain.contracts.filecoinPay.address,
        functionName: 'deposit',
        args: [token, options.to, options.amount],
    };
}
export function extractDepositEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.filecoinPay,
        logs,
        eventName: 'DepositRecorded',
        strict: true,
    });
    if (!log)
        throw new Error('`DepositRecorded` event not found.');
    return log;
}
