// @filoz/synapse-core/dist/src/pay/get-rails-for-payee-and-token.js
import { readContract } from 'viem/actions';
import { asChain } from "../chains.js";
export async function getRailsForPayeeAndToken(client, options) {
    const data = await readContract(client, getRailsForPayeeAndTokenCall({
        chain: client.chain,
        payee: options.payee,
        token: options.token,
        offset: options.offset,
        limit: options.limit,
        contractAddress: options.contractAddress,
    }));
    return parseGetRailsForPayeeAndToken(data);
}
export function getRailsForPayeeAndTokenCall(options) {
    const chain = asChain(options.chain);
    const token = options.token ?? chain.contracts.usdfc.address;
    return {
        abi: chain.contracts.filecoinPay.abi,
        address: options.contractAddress ?? chain.contracts.filecoinPay.address,
        functionName: 'getRailsForPayeeAndToken',
        args: [options.payee, token, options.offset ?? 0n, options.limit ?? 0n],
    };
}
export function parseGetRailsForPayeeAndToken(data) {
    return {
        results: data[0].map((rail) => ({
            railId: rail.railId,
            isTerminated: rail.isTerminated,
            endEpoch: rail.endEpoch,
        })),
        nextOffset: data[1],
        total: data[2],
    };
}
