// @filoz/synapse-core/dist/src/pay/operator-approvals.js
import { readContract } from 'viem/actions';
import { asChain } from "../chains.js";
export async function operatorApprovals(client, options) {
    const data = await readContract(client, operatorApprovalsCall({
        chain: client.chain,
        token: options.token,
        address: options.address,
        operator: options.operator,
        contractAddress: options.contractAddress,
    }));
    return parseOperatorApprovals(data);
}
export function operatorApprovalsCall(options) {
    const chain = asChain(options.chain);
    const token = options.token ?? chain.contracts.usdfc.address;
    const operator = options.operator ?? chain.contracts.fwss.address;
    return {
        abi: chain.contracts.filecoinPay.abi,
        address: options.contractAddress ?? chain.contracts.filecoinPay.address,
        functionName: 'operatorApprovals',
        args: [token, options.address, operator],
    };
}
export function parseOperatorApprovals(data) {
    return {
        isApproved: data[0],
        rateAllowance: data[1],
        lockupAllowance: data[2],
        rateUsage: data[3],
        lockupUsage: data[4],
        maxLockupPeriod: data[5],
    };
}
