// @filoz/synapse-core/dist/src/pay/payments.js
import { maxUint256, parseSignature } from 'viem';
import { simulateContract, writeContract, } from 'viem/actions';
import { getChain } from "../chains.js";
import * as erc20 from "../erc20/index.js";
import { ValidationError } from "../errors/base.js";
import { DepositAmountError, InsufficientBalanceError } from "../errors/pay.js";
import { signErc20Permit } from "../typed-data/sign-erc20-permit.js";
import { LOCKUP_PERIOD, TIME_CONSTANTS } from "../utils/constants.js";
export async function depositAndApprove(client, options) {
    const chain = getChain(client.chain.id);
    const token = options.token ?? chain.contracts.usdfc.address;
    const operator = options.operator ?? chain.contracts.fwss.address;
    const address = options.address ?? client.account.address;
    const spender = options.spender ?? chain.contracts.filecoinPay.address;
    const rateAllowance = options.rateAllowance ?? maxUint256;
    const lockupAllowance = options.lockupAllowance ?? maxUint256;
    const maxLockupPeriod = options.maxLockupPeriod ?? LOCKUP_PERIOD;
    if (rateAllowance < 0n || lockupAllowance < 0n || maxLockupPeriod < 0n) {
        throw new ValidationError('Allowance or lockup period values cannot be negative');
    }
    if (options.amount <= 0n) {
        throw new DepositAmountError(options.amount);
    }
    const { value: balance, name, nonce, version, } = await erc20.balanceForPermit(client, {
        address: address,
        token: token,
    });
    if (balance < options.amount) {
        throw new InsufficientBalanceError(balance, options.amount);
    }
    const deadline = options.deadline ?? BigInt(Math.floor(Date.now() / 1000) + TIME_CONSTANTS.PERMIT_DEADLINE_DURATION);
    const structuredSignature = parseSignature(await signErc20Permit(client, {
        amount: options.amount,
        nonce,
        deadline,
        name,
        version,
        token,
        spender,
    }));
    const { request } = await simulateContract(client, {
        account: client.account,
        address: chain.contracts.filecoinPay.address,
        abi: chain.contracts.filecoinPay.abi,
        functionName: 'depositWithPermitAndApproveOperator',
        args: [
            token,
            address,
            options.amount,
            deadline,
            Number(structuredSignature.v),
            structuredSignature.r,
            structuredSignature.s,
            operator,
            rateAllowance,
            lockupAllowance,
            maxLockupPeriod,
        ],
    });
    const hash = await writeContract(client, request);
    return hash;
}
