// @filoz/synapse-core/dist/src/pay/set-operator-approval.js
import { maxUint256, parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
import { ValidationError } from "../errors/base.js";
import { LOCKUP_PERIOD } from "../utils/constants.js";
export async function setOperatorApproval(client, options) {
    const { request } = await simulateContract(client, setOperatorApprovalCall({
        chain: client.chain,
        token: options.token,
        operator: options.operator,
        approve: options.approve,
        rateAllowance: options.rateAllowance,
        lockupAllowance: options.lockupAllowance,
        maxLockupPeriod: options.maxLockupPeriod,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function setOperatorApprovalSync(client, options) {
    const hash = await setOperatorApproval(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractSetOperatorApprovalEvent(receipt.logs);
    return { receipt, event };
}
export function setOperatorApprovalCall(options) {
    const chain = asChain(options.chain);
    const token = options.token ?? chain.contracts.usdfc.address;
    const operator = options.operator ?? chain.contracts.fwss.address;
    const rateAllowance = options.rateAllowance ?? (options.approve ? maxUint256 : 0n);
    const lockupAllowance = options.lockupAllowance ?? (options.approve ? maxUint256 : 0n);
    const maxLockupPeriod = options.maxLockupPeriod ?? (options.approve ? LOCKUP_PERIOD : 0n);
    if (rateAllowance < 0n || lockupAllowance < 0n || maxLockupPeriod < 0n) {
        throw new ValidationError('Allowance or lockup period values cannot be negative');
    }
    return {
        abi: chain.contracts.filecoinPay.abi,
        address: options.contractAddress ?? chain.contracts.filecoinPay.address,
        functionName: 'setOperatorApproval',
        args: [token, operator, options.approve, rateAllowance, lockupAllowance, maxLockupPeriod],
    };
}
export function extractSetOperatorApprovalEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.filecoinPay,
        logs,
        eventName: 'OperatorApprovalUpdated',
        strict: true,
    });
    if (!log)
        throw new Error('`OperatorApprovalUpdated` event not found.');
    return log;
}
