// @filoz/synapse-core/dist/src/pay/settle-terminated-rail-without-validation.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
export async function settleTerminatedRailWithoutValidation(client, options) {
    const { request } = await simulateContract(client, settleTerminatedRailWithoutValidationCall({
        chain: client.chain,
        railId: options.railId,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function settleTerminatedRailWithoutValidationSync(client, options) {
    const hash = await settleTerminatedRailWithoutValidation(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractSettleTerminatedRailWithoutValidationEvent(receipt.logs);
    return { receipt, event };
}
export function settleTerminatedRailWithoutValidationCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.filecoinPay.abi,
        address: options.contractAddress ?? chain.contracts.filecoinPay.address,
        functionName: 'settleTerminatedRailWithoutValidation',
        args: [options.railId],
    };
}
export function extractSettleTerminatedRailWithoutValidationEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.filecoinPay,
        logs,
        eventName: 'RailSettled',
        strict: true,
    });
    if (!log)
        throw new Error('`RailSettled` event not found.');
    return log;
}
