// @filoz/synapse-core/dist/src/pay/settle-rail.js
import { parseEventLogs } from 'viem';
import { getBlockNumber, simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
export async function settleRail(client, options) {
    const untilEpoch = options.untilEpoch ??
        (await getBlockNumber(client, {
            cacheTime: 0,
        }));
    const { request } = await simulateContract(client, settleRailCall({
        chain: client.chain,
        railId: options.railId,
        untilEpoch,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function settleRailSync(client, options) {
    const hash = await settleRail(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractSettleRailEvent(receipt.logs);
    return { receipt, event };
}
export function settleRailCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.filecoinPay.abi,
        address: options.contractAddress ?? chain.contracts.filecoinPay.address,
        functionName: 'settleRail',
        args: [options.railId, options.untilEpoch],
    };
}
export function extractSettleRailEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.filecoinPay,
        logs,
        eventName: 'RailSettled',
        strict: true,
    });
    if (!log)
        throw new Error('`RailSettled` event not found.');
    return log;
}
