// @filoz/synapse-core/dist/src/pay/withdraw.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
import { ValidationError } from "../errors/base.js";
import { InsufficientAvailableFundsError } from "../errors/pay.js";
import { accounts } from "./accounts.js";
export async function withdraw(client, options) {
    if (options.amount <= 0n) {
        throw new ValidationError('Withdraw amount must be greater than 0');
    }
    const account = await accounts(client, {
        address: client.account.address,
        token: options.token,
        contractAddress: options.contractAddress,
    });
    if (account.availableFunds < options.amount) {
        throw new InsufficientAvailableFundsError(account.availableFunds, options.amount);
    }
    const { request } = await simulateContract(client, withdrawCall({
        chain: client.chain,
        token: options.token,
        amount: options.amount,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function withdrawSync(client, options) {
    const hash = await withdraw(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractWithdrawEvent(receipt.logs);
    return { receipt, event };
}
export function withdrawCall(options) {
    const chain = asChain(options.chain);
    const token = options.token ?? chain.contracts.usdfc.address;
    if (options.amount <= 0n) {
        throw new ValidationError('Withdraw amount must be greater than 0');
    }
    return {
        abi: chain.contracts.filecoinPay.abi,
        address: options.contractAddress ?? chain.contracts.filecoinPay.address,
        functionName: 'withdraw',
        args: [token, options.amount],
    };
}
export function extractWithdrawEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.filecoinPay,
        logs,
        eventName: 'WithdrawRecorded',
        strict: true,
    });
    if (!log)
        throw new Error('`WithdrawRecorded` event not found.');
    return log;
}
