// @filoz/synapse-core/dist/src/session-key/actions.js
import { simulateContract, writeContract } from 'viem/actions';
import { asChain } from "../chains.js";
import { authorizationExpiry } from "./authorization-expiry.js";
import { SESSION_KEY_PERMISSIONS } from "./permissions.js";
export async function isExpired(client, options) {
    const expiry = await authorizationExpiry(client, options);
    return expiry < BigInt(Math.floor(Date.now() / 1000));
}
export async function login(client, options) {
    const chain = asChain(client.chain);
    const expiresAt = BigInt(Math.floor(Date.now() / 1000) + 3600);
    const { request } = await simulateContract(client, {
        address: chain.contracts.sessionKeyRegistry.address,
        abi: chain.contracts.sessionKeyRegistry.abi,
        functionName: 'login',
        args: [
            options.address,
            options.expiresAt ?? expiresAt,
            [...new Set(options.permissions)].map((permission) => SESSION_KEY_PERMISSIONS[permission]),
            options.origin ?? 'synapse',
        ],
    });
    const hash = await writeContract(client, request);
    return hash;
}
export async function revoke(client, options) {
    const chain = asChain(client.chain);
    const { request } = await simulateContract(client, {
        address: chain.contracts.sessionKeyRegistry.address,
        abi: chain.contracts.sessionKeyRegistry.abi,
        functionName: 'revoke',
        args: [
            options.address,
            [...new Set(options.permissions)].map((permission) => SESSION_KEY_PERMISSIONS[permission]),
            options.origin ?? 'synapse',
        ],
    });
    const hash = await writeContract(client, request);
    return hash;
}
