// @filoz/synapse-core/dist/src/session-key/permissions.js
import { TypedData } from 'ox';
import { keccak256, stringToHex } from 'viem';
import { EIP712Types } from "../typed-data/type-definitions.js";
function typeHash(type) {
    return keccak256(stringToHex(TypedData.encodeType(type)));
}
export const ALL_PERMISSIONS = [
    'CreateDataSet',
    'AddPieces',
    'SchedulePieceRemovals',
    'DeleteDataSet',
];
export const SESSION_KEY_PERMISSIONS = {
    CreateDataSet: typeHash({
        types: EIP712Types,
        primaryType: 'CreateDataSet',
    }),
    AddPieces: typeHash({
        types: EIP712Types,
        primaryType: 'AddPieces',
    }),
    SchedulePieceRemovals: typeHash({
        types: EIP712Types,
        primaryType: 'SchedulePieceRemovals',
    }),
    DeleteDataSet: typeHash({
        types: EIP712Types,
        primaryType: 'DeleteDataSet',
    }),
};
export function getPermissionFromTypeHash(typeHash) {
    for (const permission of Object.entries(SESSION_KEY_PERMISSIONS)) {
        if (permission[1] === typeHash) {
            return permission[0];
        }
    }
    throw new Error(`Permission not found for type hash: ${typeHash}`);
}
