// @filoz/synapse-core/dist/src/session-key/secp256k1.js
import { TypedEventTarget } from 'iso-web/event-target';
import { createWalletClient, } from 'viem';
import { generatePrivateKey, privateKeyToAccount } from 'viem/accounts';
import { waitForTransactionReceipt } from 'viem/actions';
import { transportFromTransportConfig } from "../utils/viem.js";
import { isExpired, login } from "./actions.js";
export class Secp256k1Key extends TypedEventTarget {
    privateKey;
    permissions;
    expiresAt;
    type;
    account;
    isConnecting = false;
    isConnected = false;
    connectPromise;
    constructor(props) {
        super();
        this.privateKey = props.privateKey;
        this.expiresAt = props.expiresAt;
        this.type = 'secp256k1';
        this.permissions = props.permissions;
        this.account = privateKeyToAccount(this.privateKey);
    }
    static create(options) {
        const key = options?.privateKey ?? generatePrivateKey();
        return new Secp256k1Key({
            privateKey: key,
            expiresAt: options?.expiresAt,
            permissions: options?.permissions ?? ['CreateDataSet', 'AddPieces', 'SchedulePieceRemovals', 'DeleteDataSet'],
        });
    }
    get connecting() {
        return this.isConnecting;
    }
    get connected() {
        return this.isConnected;
    }
    async connect(client) {
        if (this.isConnecting) {
            throw new Error('Already connecting');
        }
        this.isConnecting = true;
        try {
            const _isExpired = await this.isValid(client, this.permissions[0]);
            if (_isExpired) {
                const hash = await this.refresh(client);
                this.connectPromise = waitForTransactionReceipt(client, { hash }).then((receipt) => {
                    this.isConnected = true;
                    this.emit('connected', this.account);
                    this.connectPromise = undefined;
                    return receipt;
                }, (error) => {
                    this.connectPromise = undefined;
                    this.emit('error', new Error('Failed to wait for connect', { cause: error }));
                    return undefined;
                });
            }
        }
        catch (error) {
            throw new Error('Failed to connect', { cause: error });
        }
        finally {
            this.isConnecting = false;
        }
    }
    disconnect() {
        this.isConnected = false;
        this.emit('disconnected');
        this.connectPromise = undefined;
        return Promise.resolve();
    }
    async refresh(client) {
        const hash = await login(client, {
            address: this.account.address,
            permissions: this.permissions,
            expiresAt: this.expiresAt ? BigInt(this.expiresAt) : undefined,
        });
        return hash;
    }
    async isValid(client, permission) {
        if (!this.permissions.includes(permission)) {
            return false;
        }
        if (this.connectPromise) {
            await this.connectPromise;
        }
        return isExpired(client, {
            address: client.account.address,
            sessionKeyAddress: this.account.address,
            permission: permission,
        });
    }
    client(chain, transportConfig) {
        if (!this.connected) {
            throw new Error('Not connected');
        }
        return createWalletClient({
            chain,
            transport: transportFromTransportConfig({ transportConfig }),
            account: this.account,
        });
    }
}
