// @filoz/synapse-core/src/pdp-verifier/get-scheduled-removals.ts
import type { Simplify } from 'type-fest'
import type {
  Address,
  Chain,
  Client,
  ContractFunctionParameters,
  ContractFunctionReturnType,
  ReadContractErrorType,
  Transport,
} from 'viem'
import { readContract } from 'viem/actions'
import type { pdpVerifierAbi } from '../abis/generated.ts'
import { asChain } from '../chains.ts'
import type { ActionCallChain } from '../types.ts'

export namespace getScheduledRemovals {
  export type OptionsType = {
    /** The ID of the data set to get the scheduled removals for. */
    dataSetId: bigint
    /** PDP Verifier contract address. If not provided, the default is the PDP Verifier contract address for the chain. */
    contractAddress?: Address
  }

  export type OutputType = ContractFunctionReturnType<typeof pdpVerifierAbi, 'pure' | 'view', 'getScheduledRemovals'>

  export type ErrorType = asChain.ErrorType | ReadContractErrorType
}

/**
 * Get the scheduled removals for a data set (deduped)
 *
 * @example
 * ```ts
 * import { getScheduledRemovals } from '@filoz/synapse-core/pdp-verifier'
 * import { calibration } from '@filoz/synapse-core/chains'
 * import { createPublicClient, http } from 'viem'
 *
 * const client = createPublicClient({
 *   chain: calibration,
 *   transport: http(),
 * })
 *
 * const pieceIds = await getScheduledRemovals(client, {
 *   dataSetId: 1n,
 * })
 * ```
 *
 * @param client - The client to use to get the scheduled removals.
 * @param options - {@link getScheduledRemovals.OptionsType}
 * @returns The piece IDs scheduled for removal {@link getScheduledRemovals.OutputType}
 * @throws Errors {@link getScheduledRemovals.ErrorType}
 */
export async function getScheduledRemovals(
  client: Client<Transport, Chain>,
  options: getScheduledRemovals.OptionsType
): Promise<getScheduledRemovals.OutputType> {
  const data = await readContract(
    client,
    getScheduledRemovalsCall({
      chain: client.chain,
      dataSetId: options.dataSetId,
      contractAddress: options.contractAddress,
    })
  )
  return Array.from(new Set(data))
}

export namespace getScheduledRemovalsCall {
  export type OptionsType = Simplify<getScheduledRemovals.OptionsType & ActionCallChain>
  export type ErrorType = asChain.ErrorType
  export type OutputType = ContractFunctionParameters<typeof pdpVerifierAbi, 'pure' | 'view', 'getScheduledRemovals'>
}

/**
 * Create a call to the getScheduledRemovals function
 *
 * This function is used to create a call to the getScheduledRemovals function for use with the multicall or readContract function.
 *
 * May require manual deduplication of the output if the data set has multiple scheduled removals for the same piece.
 *
 * @example
 * ```ts
 * import { getScheduledRemovalsCall } from '@filoz/synapse-core/pdp-verifier'
 * import { calibration } from '@filoz/synapse-core/chains'
 * import { createPublicClient, http } from 'viem'
 * import { multicall } from 'viem/actions'
 *
 * const client = createPublicClient({
 *   chain: calibration,
 *   transport: http(),
 * })
 *
 * const results = await multicall(client, {
 *   contracts: [
 *     getScheduledRemovalsCall({ chain: calibration, dataSetId: 1n }),
 *     getScheduledRemovalsCall({ chain: calibration, dataSetId: 2n }),
 *   ],
 * })
 * ```
 *
 * @param options - {@link getScheduledRemovalsCall.OptionsType}
 * @returns The call to the getScheduledRemovals function {@link getScheduledRemovalsCall.OutputType}
 * @throws Errors {@link getScheduledRemovalsCall.ErrorType}
 */
export function getScheduledRemovalsCall(options: getScheduledRemovalsCall.OptionsType) {
  const chain = asChain(options.chain)
  return {
    abi: chain.contracts.pdp.abi,
    address: options.contractAddress ?? chain.contracts.pdp.address,
    functionName: 'getScheduledRemovals',
    args: [options.dataSetId],
  } satisfies getScheduledRemovalsCall.OutputType
}
