// @filoz/synapse-core/dist/src/piece.js
import * as Hasher from '@web3-storage/data-segment/multihash';
import { Unpadded } from '@web3-storage/data-segment/piece/size';
import { CID } from 'multiformats/cid';
import * as Raw from 'multiformats/codecs/raw';
import * as Link from 'multiformats/link';
import { hexToBytes } from 'viem';
import { DownloadPieceError } from "./errors/pdp.js";
function parsePieceCID(pieceCidString) {
    try {
        const cid = CID.parse(pieceCidString);
        if (isValidPieceCID(cid)) {
            return cid;
        }
    }
    catch {
    }
    return null;
}
function isCID(value) {
    return typeof value === 'object' && value !== null && CID.asCID(value) !== null;
}
function isValidPieceCID(cid) {
    return cid.code === Raw.code && cid.multihash.code === Hasher.code;
}
export function asPieceCID(pieceCidInput) {
    if (pieceCidInput === null || pieceCidInput === undefined) {
        return null;
    }
    if (typeof pieceCidInput === 'string') {
        return parsePieceCID(pieceCidInput);
    }
    if (isCID(pieceCidInput)) {
        if (isValidPieceCID(pieceCidInput)) {
            return pieceCidInput;
        }
    }
    return null;
}
export function getSize(pieceCid) {
    const digest = Hasher.Digest.fromBytes(pieceCid.multihash.bytes);
    const height = digest.height;
    const padding = digest.padding;
    const rawSize = Unpadded.fromPiece({ height, padding });
    if (rawSize > Number.MAX_SAFE_INTEGER) {
        throw new Error(`Raw size ${rawSize} exceeds maximum safe integer`);
    }
    return Number(rawSize);
}
export function getSizeFromPieceCID(pieceCidInput) {
    const pieceCid = asPieceCID(pieceCidInput);
    if (pieceCid == null) {
        throw new Error('Invalid PieceCID: input must be a valid PieceCIDv2');
    }
    return getSize(pieceCid);
}
export function parse(pieceCid) {
    try {
        const cid = CID.parse(pieceCid).toV1();
        if (!isPieceCID(cid)) {
            throw new Error('Invalid PieceCID: input must be a valid PieceCIDv2');
        }
        return cid;
    }
    catch {
        throw new Error(`Invalid CID string: ${pieceCid}`);
    }
}
export function isPieceCID(cid) {
    return (typeof cid === 'object' && CID.asCID(cid) != null && cid.code === Raw.code && cid.multihash.code === Hasher.code);
}
export function calculate(data) {
    const hasher = Hasher.create();
    const chunkSize = 2048;
    for (let i = 0; i < data.length; i += chunkSize) {
        hasher.write(data.subarray(i, i + chunkSize));
    }
    const digest = hasher.digest();
    return Link.create(Raw.code, digest);
}
export async function calculateFromIterable(data) {
    const hasher = Hasher.create();
    for await (const chunk of data) {
        hasher.write(chunk);
    }
    const digest = hasher.digest();
    return Link.create(Raw.code, digest);
}
export function createPieceCIDStream() {
    const hasher = Hasher.create();
    let finished = false;
    let pieceCid = null;
    const stream = new TransformStream({
        transform(chunk, controller) {
            hasher.write(chunk);
            controller.enqueue(chunk);
        },
        flush() {
            const digest = hasher.digest();
            pieceCid = Link.create(Raw.code, digest);
            finished = true;
        },
    });
    return {
        stream,
        getPieceCID: () => {
            if (!finished) {
                return null;
            }
            return pieceCid;
        },
    };
}
export function hexToPieceCID(pieceCidHex) {
    const pieceDataBytes = hexToBytes(pieceCidHex);
    const possiblePieceCID = CID.decode(pieceDataBytes);
    const isValid = isValidPieceCID(possiblePieceCID);
    if (!isValid) {
        throw new Error(`Hex string '${pieceCidHex}' is a valid CID but not a valid PieceCID`);
    }
    return possiblePieceCID;
}
export async function downloadAndValidate(response, expectedPieceCid) {
    const parsedPieceCid = asPieceCID(expectedPieceCid);
    if (parsedPieceCid == null) {
        throw new DownloadPieceError(`Invalid PieceCID: ${String(expectedPieceCid)}`);
    }
    if (!response.ok) {
        throw new DownloadPieceError(`Download failed: ${response.status} ${response.statusText}`);
    }
    if (response.body == null) {
        throw new DownloadPieceError('Response body is null');
    }
    const { stream: pieceCidStream, getPieceCID } = createPieceCIDStream();
    const chunks = [];
    const collectStream = new TransformStream({
        transform(chunk, controller) {
            chunks.push(chunk);
            controller.enqueue(chunk);
        },
    });
    const pipelineStream = response.body.pipeThrough(pieceCidStream).pipeThrough(collectStream);
    const reader = pipelineStream.getReader();
    try {
        while (true) {
            const { done } = await reader.read();
            if (done)
                break;
        }
    }
    finally {
        reader.releaseLock();
    }
    if (chunks.length === 0) {
        throw new DownloadPieceError('Response body is empty');
    }
    const calculatedPieceCid = getPieceCID();
    if (calculatedPieceCid == null) {
        throw new DownloadPieceError('Failed to calculate PieceCID from stream');
    }
    if (calculatedPieceCid.toString() !== parsedPieceCid.toString()) {
        throw new DownloadPieceError(`PieceCID verification failed. Expected: ${String(parsedPieceCid)}, Got: ${String(calculatedPieceCid)}`);
    }
    const totalLength = chunks.reduce((acc, chunk) => acc + chunk.length, 0);
    const result = new Uint8Array(totalLength);
    let offset = 0;
    for (const chunk of chunks) {
        result.set(chunk, offset);
        offset += chunk.length;
    }
    return result;
}
