// @filoz/synapse-core/dist/src/mocks/pdp.js
import assert from 'assert';
import { HttpResponse, http } from 'msw';
import { decodeAbiParameters } from 'viem';
export function createAndAddPiecesHandler(txHash, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.post(`${baseUrl}/pdp/data-sets/create-and-add`, () => {
        return new HttpResponse(null, {
            status: 201,
            headers: { Location: `/pdp/data-sets/created/${txHash}` },
        });
    });
}
export function dataSetCreationStatusHandler(txHash, response, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.get(`${baseUrl}/pdp/data-sets/created/:txHash`, ({ params }) => {
        if (params.txHash !== txHash) {
            return new HttpResponse(null, { status: 404 });
        }
        return HttpResponse.json(response, { status: 200 });
    });
}
export function pieceAdditionStatusHandler(dataSetId, txHash, response, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.get(`${baseUrl}/pdp/data-sets/:id/pieces/added/:txHash`, ({ params }) => {
        if (params.id !== dataSetId.toString() || params.txHash !== txHash) {
            return new HttpResponse(null, { status: 404 });
        }
        return HttpResponse.json(response, { status: 200 });
    });
}
export function findPieceHandler(pieceCid, found, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.get(`${baseUrl}/pdp/piece`, ({ request }) => {
        const url = new URL(request.url);
        const queryCid = url.searchParams.get('pieceCid');
        if (queryCid !== pieceCid) {
            return HttpResponse.text(null, { status: 404 });
        }
        if (!found) {
            return HttpResponse.text(null, { status: 404 });
        }
        return HttpResponse.json({ pieceCid }, { status: 200 });
    });
}
export function findAnyPieceHandler(found, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.get(`${baseUrl}/pdp/piece`, ({ request }) => {
        const url = new URL(request.url);
        const queryCid = url.searchParams.get('pieceCid');
        if (found) {
            return HttpResponse.json({ pieceCid: queryCid });
        }
        else {
            return HttpResponse.text(null, { status: 404 });
        }
    });
}
export function postPieceHandler(pieceCid, uuid, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.post(`${baseUrl}/pdp/piece`, async ({ request }) => {
        const body = await request.json();
        assert(body != null, 'Body should be defined');
        assert(body.pieceCid != null, 'PieceCID should be defined');
        assert.strictEqual(body.pieceCid, pieceCid, 'PieceCID should match expected value');
        if (uuid == null) {
            return HttpResponse.json({
                pieceCid,
            });
        }
        return HttpResponse.text('Created', {
            status: 201,
            headers: {
                Location: `/pdp/piece/upload/${uuid}`,
            },
        });
    });
}
export function uploadPieceHandler(uuid, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.put(`${baseUrl}/pdp/piece/upload/${uuid}`, async () => {
        return HttpResponse.text('No Content', {
            status: 204,
        });
    });
}
export function postPieceUploadsHandler(uuid, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.post(`${baseUrl}/pdp/piece/uploads`, async () => {
        return HttpResponse.text('Created', {
            status: 201,
            headers: {
                Location: `/pdp/piece/uploads/${uuid}`,
            },
        });
    });
}
export function uploadPieceStreamingHandler(uuid, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.put(`${baseUrl}/pdp/piece/uploads/${uuid}`, async ({ request }) => {
        await request.arrayBuffer();
        return HttpResponse.text('No Content', {
            status: 204,
        });
    });
}
export function finalizePieceUploadHandler(uuid, expectedPieceCid, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.post(`${baseUrl}/pdp/piece/uploads/${uuid}`, async ({ request }) => {
        const body = await request.json();
        if (expectedPieceCid != null) {
            assert.equal(body.pieceCid, expectedPieceCid, 'PieceCID should match expected value');
        }
        return HttpResponse.json({ pieceCid: body.pieceCid }, {
            status: 200,
        });
    });
}
export function streamingUploadHandlers(options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    let uploadCounter = 0;
    return [
        http.post(`${baseUrl}/pdp/piece/uploads`, async () => {
            const uuid = `${(++uploadCounter).toString(16).padStart(8, '0')}-0000-0000-0000-000000000000`;
            return HttpResponse.text('Created', {
                status: 201,
                headers: {
                    Location: `/pdp/piece/uploads/${uuid}`,
                },
            });
        }),
        http.put(`${baseUrl}/pdp/piece/uploads/:uuid`, async ({ request }) => {
            await request.arrayBuffer();
            return HttpResponse.text('No Content', {
                status: 204,
            });
        }),
        http.post(`${baseUrl}/pdp/piece/uploads/:uuid`, async ({ request }) => {
            const body = await request.json();
            return HttpResponse.json({ pieceCid: body.pieceCid }, {
                status: 200,
            });
        }),
    ];
}
export function decodeMetadataFromCreateDataSetExtraData(extraData) {
    const decoded = decodeAbiParameters([{ type: 'address' }, { type: 'uint256' }, { type: 'string[]' }, { type: 'string[]' }, { type: 'bytes' }], extraData);
    return {
        keys: decoded[2],
        values: decoded[3],
    };
}
export function decodePieceMetadataFromExtraData(extraData) {
    const decoded = decodeAbiParameters([{ type: 'uint256' }, { type: 'string[][]' }, { type: 'string[][]' }, { type: 'bytes' }], extraData);
    return {
        keys: decoded[1],
        values: decoded[2],
    };
}
export function createDataSetWithMetadataCapture(txHash, captureCallback, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.post(`${baseUrl}/pdp/data-sets`, async ({ request }) => {
        const body = (await request.json());
        if (!body.extraData) {
            return new HttpResponse(JSON.stringify({ error: 'Missing extraData' }), { status: 400 });
        }
        try {
            const metadata = decodeMetadataFromCreateDataSetExtraData(body.extraData);
            captureCallback(metadata);
            if (options.debug) {
                console.debug('PDP Mock: captured metadata', metadata);
            }
        }
        catch (error) {
            if (options.debug) {
                console.debug('PDP Mock: failed to decode extraData', error);
            }
        }
        return new HttpResponse(null, {
            status: 201,
            headers: { Location: `/pdp/data-sets/created/${txHash}` },
        });
    });
}
export function addPiecesWithMetadataCapture(dataSetId, txHash, captureCallback, options = {}) {
    const baseUrl = options.baseUrl ?? 'https://pdp.example.com';
    return http.post(`${baseUrl}/pdp/data-sets/:id/pieces`, async ({ params, request }) => {
        if (params.id !== dataSetId.toString()) {
            return new HttpResponse(null, { status: 404 });
        }
        const body = await request.json();
        if (!body.extraData) {
            return new HttpResponse(JSON.stringify({ error: 'Missing extraData' }), { status: 400 });
        }
        try {
            const metadata = decodePieceMetadataFromExtraData(body.extraData);
            captureCallback(metadata);
            if (options.debug) {
                console.debug('PDP Mock: captured piece metadata', metadata);
            }
        }
        catch (error) {
            if (options.debug) {
                console.debug('PDP Mock: failed to decode extraData', error);
            }
        }
        return new HttpResponse(null, {
            status: 201,
            headers: { Location: `/pdp/data-sets/${dataSetId}/pieces/added/${txHash}` },
        });
    });
}
