// @filoz/synapse-sdk/dist/src/synapse.js
import { asChain } from '@filoz/synapse-core/chains';
import { createClient, http, isAddress, publicActions, } from 'viem';
import { FilBeamService } from "./filbeam/index.js";
import { PaymentsService } from "./payments/index.js";
import { ChainRetriever, FilBeamRetriever } from "./retriever/index.js";
import { SPRegistryService } from "./sp-registry/index.js";
import { StorageManager } from "./storage/manager.js";
import { DEFAULT_CHAIN } from "./utils/constants.js";
import { WarmStorageService } from "./warm-storage/index.js";
export class Synapse {
    _withCDN;
    _payments;
    _warmStorageService;
    _storageManager;
    _filbeamService;
    _providers;
    _client;
    _chain;
    static create(options) {
        const client = createClient({
            chain: options.chain ?? DEFAULT_CHAIN,
            transport: options.transport ?? http(),
            account: options.account,
            name: 'Synapse Client',
            key: 'synapse-client',
        });
        if (client.account.type === 'json-rpc' && client.transport.type !== 'custom') {
            throw new Error('Transport must be a custom transport. See https://viem.sh/docs/clients/transports/custom.');
        }
        return new Synapse({ client, withCDN: options.withCDN });
    }
    constructor(options) {
        this._client = options.client.extend(publicActions);
        this._chain = asChain(options.client.chain);
        this._withCDN = options.withCDN ?? false;
        this._providers = new SPRegistryService({ client: options.client });
        this._filbeamService = new FilBeamService(this._chain);
        this._warmStorageService = new WarmStorageService({ client: options.client });
        this._payments = new PaymentsService({ client: options.client });
        this._storageManager = new StorageManager({
            synapse: this,
            warmStorageService: this._warmStorageService,
            pieceRetriever: new FilBeamRetriever({
                baseRetriever: new ChainRetriever({
                    warmStorageService: this._warmStorageService,
                    spRegistry: this._providers,
                }),
                chain: this._chain,
            }),
            withCDN: this._withCDN,
        });
    }
    get client() {
        return this._client;
    }
    get chain() {
        return this._chain;
    }
    get payments() {
        return this._payments;
    }
    get storage() {
        return this._storageManager;
    }
    get filbeam() {
        return this._filbeamService;
    }
    get providers() {
        return this._providers;
    }
    async getProviderInfo(providerAddress) {
        try {
            if (typeof providerAddress === 'string') {
                try {
                    isAddress(providerAddress);
                }
                catch {
                    throw new Error(`Invalid provider address: ${providerAddress}`);
                }
            }
            let providerInfo;
            if (typeof providerAddress === 'string') {
                providerInfo = await this._providers.getProviderByAddress({ address: providerAddress });
            }
            else {
                providerInfo = await this._providers.getProvider({ providerId: providerAddress });
            }
            if (providerInfo == null) {
                throw new Error(`Provider ${providerAddress} not found in registry`);
            }
            return providerInfo;
        }
        catch (error) {
            if (error instanceof Error && error.message.includes('Invalid provider address')) {
                throw error;
            }
            if (error instanceof Error && error.message.includes('not found')) {
                throw error;
            }
            throw new Error(`Failed to get provider info: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
}
