// @filoz/synapse-sdk/dist/src/payments/service.js
import { asChain } from '@filoz/synapse-core/chains';
import * as ERC20 from '@filoz/synapse-core/erc20';
import * as Pay from '@filoz/synapse-core/pay';
import { signErc20Permit } from '@filoz/synapse-core/typed-data';
import { createClient, http, parseSignature, } from 'viem';
import { getBalance, getBlockNumber, simulateContract, writeContract } from 'viem/actions';
import { createError, DEFAULT_CHAIN, TIME_CONSTANTS, TOKENS } from "../utils/index.js";
export class PaymentsService {
    _client;
    constructor(options) {
        this._client = options.client;
    }
    static create(options) {
        const client = createClient({
            chain: options.chain ?? DEFAULT_CHAIN,
            transport: options.transport ?? http(),
            account: options.account,
            name: 'PaymentsService',
            key: 'payments-service',
        });
        if (client.account.type === 'json-rpc' && client.transport.type !== 'custom') {
            throw new Error('Transport must be a custom transport. See https://viem.sh/docs/clients/transports/custom.');
        }
        return new PaymentsService({ client });
    }
    async balance(options = { token: TOKENS.USDFC }) {
        if (options.token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'payments contract balance check', `Token "${options.token}" is not supported. Currently only USDFC token is supported for payments contract balance queries.`);
        }
        const accountInfo = await this.accountInfo({ token: options.token });
        return accountInfo.availableFunds;
    }
    async accountInfo(options = { token: TOKENS.USDFC }) {
        if (options.token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'account info', `Token "${options.token}" is not supported. Currently only USDFC token is supported.`);
        }
        return await Pay.accounts(this._client, {
            address: this._client.account.address,
        });
    }
    async walletBalance(options = {}) {
        const { token = TOKENS.FIL } = options;
        if (token === TOKENS.FIL) {
            try {
                const balance = await getBalance(this._client, {
                    address: this._client.account.address,
                });
                return balance;
            }
            catch (error) {
                throw createError('PaymentsService', 'wallet FIL balance check', 'Unable to retrieve FIL balance from wallet. This could be due to network connectivity issues, RPC endpoint problems, or wallet connection issues.', error);
            }
        }
        if (token === TOKENS.USDFC) {
            try {
                const balance = await ERC20.balance(this._client, {
                    address: this._client.account.address,
                });
                return balance.value;
            }
            catch (error) {
                throw createError('PaymentsService', 'wallet USDFC balance check', 'Unexpected error while checking USDFC token balance in wallet.', error);
            }
        }
        throw createError('PaymentsService', 'wallet balance', `Token "${token}" is not supported. Currently only FIL and USDFC tokens are supported.`);
    }
    decimals() {
        return 18;
    }
    async allowance(options) {
        const { spender, token = TOKENS.USDFC } = options;
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'allowance', `Token "${token}" is not supported. Currently only USDFC token is supported.`);
        }
        try {
            const balance = await ERC20.balance(this._client, {
                address: this._client.account.address,
                spender,
            });
            return balance.allowance;
        }
        catch (error) {
            throw createError('PaymentsService', 'allowance check', 'Failed to check token allowance. This could indicate network connectivity issues or an invalid spender address.', error);
        }
    }
    async approve(options) {
        const { spender, amount, token = TOKENS.USDFC } = options;
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'approve', `Token "${token}" is not supported. Currently only USDFC token is supported.`);
        }
        try {
            const approveTx = await ERC20.approve(this._client, {
                spender: spender,
                amount,
            });
            return approveTx;
        }
        catch (error) {
            throw createError('PaymentsService', 'approve', `Failed to approve ${spender} to spend ${amount.toString()} ${token}`, error);
        }
    }
    async approveService(options = {}) {
        const { service, rateAllowance, lockupAllowance, maxLockupPeriod, token = TOKENS.USDFC } = options;
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'approveService', `Token "${token}" is not supported. Currently only USDFC token is supported.`);
        }
        try {
            const approveTx = await Pay.setOperatorApproval(this._client, {
                operator: service,
                approve: true,
                rateAllowance: rateAllowance,
                lockupAllowance: lockupAllowance,
                maxLockupPeriod: maxLockupPeriod,
            });
            return approveTx;
        }
        catch (error) {
            throw createError('PaymentsService', 'approveService', `Failed to approve service ${service} as operator for ${token}`, error);
        }
    }
    async revokeService(options = {}) {
        const { service, token = TOKENS.USDFC } = options;
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'revokeService', `Token "${token}" is not supported. Currently only USDFC token is supported.`);
        }
        try {
            const revokeTx = await Pay.setOperatorApproval(this._client, {
                operator: service,
                approve: false,
            });
            return revokeTx;
        }
        catch (error) {
            throw createError('PaymentsService', 'revokeService', `Failed to revoke service ${service} as operator for ${token}`, error);
        }
    }
    async serviceApproval(options = {}) {
        const { service, token = TOKENS.USDFC } = options;
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'serviceApproval', `Token "${token}" is not supported. Currently only USDFC token is supported.`);
        }
        try {
            const approval = await Pay.operatorApprovals(this._client, {
                address: this._client.account.address,
                operator: service,
            });
            return approval;
        }
        catch (error) {
            throw createError('PaymentsService', 'serviceApproval', `Failed to check service approval status for ${service}`, error);
        }
    }
    async deposit(options) {
        const { amount, token = TOKENS.USDFC } = options;
        const chain = asChain(this._client.chain);
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'deposit', `Unsupported token: ${token}`);
        }
        if (amount <= 0n) {
            throw createError('PaymentsService', 'deposit', 'Invalid amount');
        }
        const erc20Balance = await ERC20.balance(this._client, {
            address: this._client.account.address,
        });
        if (erc20Balance.value < amount) {
            throw createError('PaymentsService', 'deposit', `Insufficient USDFC: have ${erc20Balance.value.toString()}, need ${amount.toString()}`);
        }
        const currentAllowance = erc20Balance.allowance;
        options?.onAllowanceCheck?.(currentAllowance, amount);
        if (currentAllowance < amount) {
            const { receipt } = await ERC20.approveSync(this._client, {
                spender: chain.contracts.filecoinPay.address,
                amount,
                onHash: options?.onApprovalTransaction,
            });
            options?.onApprovalConfirmed?.(receipt);
        }
        options?.onDepositStarting?.();
        const depositTx = await Pay.deposit(this._client, {
            amount,
            to: options?.to,
        });
        return depositTx;
    }
    async depositWithPermit(options) {
        const { amount, token = TOKENS.USDFC, deadline } = options;
        const chain = asChain(this._client.chain);
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'depositWithPermit', `Unsupported token: ${token}`);
        }
        if (amount <= 0n) {
            throw createError('PaymentsService', 'depositWithPermit', 'Invalid amount');
        }
        const permitDeadline = deadline == null ? BigInt(Math.floor(Date.now() / 1000) + TIME_CONSTANTS.PERMIT_DEADLINE_DURATION) : deadline;
        const balance = await ERC20.balanceForPermit(this._client, {
            address: this._client.account.address,
        });
        if (balance.value < amount) {
            throw createError('PaymentsService', 'depositWithPermit', 'Insufficient balance');
        }
        const signature = parseSignature(await signErc20Permit(this._client, {
            amount,
            nonce: balance.nonce,
            deadline: permitDeadline,
            name: balance.name,
            version: balance.version,
            token: chain.contracts.usdfc.address,
            spender: chain.contracts.filecoinPay.address,
        }));
        try {
            const { request } = await simulateContract(this._client, {
                account: this._client.account,
                address: chain.contracts.filecoinPay.address,
                abi: chain.contracts.filecoinPay.abi,
                functionName: 'depositWithPermit',
                args: [
                    chain.contracts.usdfc.address,
                    this._client.account.address,
                    amount,
                    permitDeadline,
                    Number(signature.v),
                    signature.r,
                    signature.s,
                ],
            });
            const hash = await writeContract(this._client, request);
            return hash;
        }
        catch (error) {
            throw createError('PaymentsService', 'depositWithPermit', 'Failed to execute depositWithPermit on Payments contract.', error);
        }
    }
    async depositWithPermitAndApproveOperator(options) {
        const { amount, operator, rateAllowance, lockupAllowance, maxLockupPeriod, deadline, token = TOKENS.USDFC, } = options;
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'depositWithPermitAndApproveOperator', `Unsupported token: ${token}`);
        }
        try {
            const hash = await Pay.depositAndApprove(this._client, {
                amount,
                operator,
                rateAllowance,
                lockupAllowance,
                maxLockupPeriod,
                deadline,
            });
            return hash;
        }
        catch (error) {
            throw createError('PaymentsService', 'depositWithPermitAndApproveOperator', 'Failed to execute depositWithPermitAndApproveOperator on Payments contract.', error);
        }
    }
    async withdraw(options) {
        const { amount, token = TOKENS.USDFC } = options;
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'withdraw', `Unsupported token: ${token}`);
        }
        return Pay.withdraw(this._client, {
            amount,
        });
    }
    async settle(options) {
        return Pay.settleRail(this._client, options);
    }
    async getSettlementAmounts(options) {
        const { railId, untilEpoch } = options;
        const _untilEpoch = untilEpoch ??
            (await getBlockNumber(this._client, {
                cacheTime: 0,
            }));
        try {
            const { result } = await simulateContract(this._client, Pay.settleRailCall({
                railId,
                untilEpoch: _untilEpoch,
                chain: this._client.chain,
            }));
            return {
                totalSettledAmount: result[0],
                totalNetPayeeAmount: result[1],
                totalOperatorCommission: result[2],
                totalNetworkFee: result[3],
                finalSettledEpoch: result[4],
                note: result[5],
            };
        }
        catch (error) {
            throw createError('PaymentsService', 'getSettlementAmounts', `Failed to get settlement amounts for rail ${railId.toString()} up to epoch ${_untilEpoch.toString()}`, error);
        }
    }
    async settleTerminatedRail(options) {
        return Pay.settleTerminatedRailWithoutValidation(this._client, options);
    }
    async getRail(options) {
        try {
            const rail = await Pay.getRail(this._client, options);
            return rail;
        }
        catch (error) {
            if (error.message?.includes('RailInactiveOrSettled')) {
                throw createError('PaymentsService', 'getRail', `Rail ${options.railId.toString()} does not exist or is inactive`);
            }
            throw createError('PaymentsService', 'getRail', `Failed to get rail ${options.railId.toString()}`, error);
        }
    }
    async settleAuto(options) {
        const rail = await this.getRail(options);
        if (rail.endEpoch > 0n) {
            return await this.settleTerminatedRail(options);
        }
        else {
            return await this.settle(options);
        }
    }
    async getRailsAsPayer(options = {}) {
        const { token = TOKENS.USDFC } = options;
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'getRailsAsPayer', `Token "${token}" is not supported. Currently only USDFC token is supported.`);
        }
        try {
            const { results } = await Pay.getRailsForPayerAndToken(this._client, {
                payer: this._client.account.address,
            });
            return results;
        }
        catch (error) {
            throw createError('PaymentsService', 'getRailsAsPayer', 'Failed to get rails where wallet is payer', error);
        }
    }
    async getRailsAsPayee(options = {}) {
        const { token = TOKENS.USDFC } = options;
        if (token !== TOKENS.USDFC) {
            throw createError('PaymentsService', 'getRailsAsPayee', `Token "${token}" is not supported. Currently only USDFC token is supported.`);
        }
        try {
            const { results } = await Pay.getRailsForPayeeAndToken(this._client, {
                payee: this._client.account.address,
            });
            return results;
        }
        catch (error) {
            throw createError('PaymentsService', 'getRailsAsPayee', 'Failed to get rails where wallet is payee', error);
        }
    }
}
