// @filoz/synapse-sdk/dist/src/retriever/chain.js
import { createError } from "../utils/index.js";
import { fetchPiecesFromProviders } from "./utils.js";
export class ChainRetriever {
    warmStorageService;
    childRetriever;
    spRegistry;
    constructor(options) {
        this.warmStorageService = options.warmStorageService;
        this.spRegistry = options.spRegistry;
        this.childRetriever = options.childRetriever;
    }
    async findProviders(options) {
        const { client, providerAddress } = options;
        if (providerAddress != null) {
            const provider = await this.spRegistry.getProviderByAddress({ address: providerAddress });
            if (provider == null) {
                throw createError('ChainRetriever', 'findProviders', `Provider ${providerAddress} not found in registry`);
            }
            return [provider];
        }
        const dataSets = await this.warmStorageService.getClientDataSetsWithDetails({ address: client });
        const validDataSets = dataSets.filter((ds) => ds.isLive && ds.activePieceCount > 0);
        if (validDataSets.length === 0) {
            throw createError('ChainRetriever', 'findProviders', `No active data sets with data found for client ${client}`);
        }
        const uniqueProviderIds = [...new Set(validDataSets.map((ds) => ds.providerId))];
        const providerInfos = await this.spRegistry.getProviders({ providerIds: uniqueProviderIds });
        const validProviderInfos = providerInfos.filter((info) => info != null);
        if (validProviderInfos.length === 0) {
            throw createError('ChainRetriever', 'findProviders', 'No valid providers found (all providers may have been removed from registry or are inactive)');
        }
        return validProviderInfos;
    }
    async fetchPiece(options) {
        const { pieceCid, client, providerAddress, withCDN, signal } = options;
        const tryChildOrThrow = async (reason, cause) => {
            if (this.childRetriever !== undefined) {
                return await this.childRetriever.fetchPiece({ pieceCid, client, providerAddress, withCDN, signal });
            }
            throw createError('ChainRetriever', 'fetchPiece', `Failed to retrieve piece ${pieceCid.toString()}: ${reason}`, cause);
        };
        let providersToTry = [];
        try {
            providersToTry = await this.findProviders({ client, providerAddress });
        }
        catch (error) {
            const message = error instanceof Error ? error.message : 'Provider discovery failed';
            return await tryChildOrThrow(message, error);
        }
        if (providersToTry.length === 0) {
            return await tryChildOrThrow('No providers found and no additional retriever method was configured');
        }
        try {
            return await fetchPiecesFromProviders(providersToTry, pieceCid, 'ChainRetriever', signal);
        }
        catch (error) {
            return await tryChildOrThrow('All provider retrieval attempts failed and no additional retriever method was configured', error);
        }
    }
}
