// @filoz/synapse-sdk/dist/src/retriever/utils.js
import * as SP from '@filoz/synapse-core/sp';
import { createPieceUrlPDP } from '@filoz/synapse-core/utils';
import { createError } from "../utils/errors.js";
export async function fetchPiecesFromProviders(providers, pieceCid, retrieverName, signal) {
    const failures = [];
    const abortControllers = [];
    const providerAttempts = providers.map(async (provider, index) => {
        const controller = new AbortController();
        abortControllers[index] = controller;
        const _signal = signal ? AbortSignal.any([controller.signal, signal]) : controller.signal;
        try {
            await SP.findPiece({
                serviceURL: provider.pdp.serviceURL,
                pieceCid,
                signal: _signal,
            });
            const downloadUrl = createPieceUrlPDP({
                cid: pieceCid.toString(),
                serviceURL: provider.pdp.serviceURL,
            });
            const response = await fetch(downloadUrl, {
                signal: _signal,
            });
            if (response.ok) {
                return { response, index };
            }
            failures.push({
                provider: provider.serviceProvider,
                error: `download returned ${response.status}`,
            });
            throw new Error(`Download failed with status ${response.status}`);
        }
        catch (error) {
            const errorMsg = error.message ?? 'Unknown error';
            if (!failures.some((f) => f.provider === provider.serviceProvider)) {
                failures.push({ provider: provider.serviceProvider, error: errorMsg });
            }
            if (errorMsg !== 'This operation was aborted') {
                console.warn(`Failed to fetch from provider ${provider.serviceProvider}:`, errorMsg);
            }
            throw error;
        }
    });
    try {
        const { response, index: winnerIndex } = await Promise.any(providerAttempts);
        abortControllers.forEach((ctrl, i) => {
            if (i !== winnerIndex) {
                ctrl.abort();
            }
        });
        return response;
    }
    catch (error) {
        if (error instanceof AggregateError) {
            const failureDetails = failures.map((f) => `${f.provider}: ${f.error}`).join('; ');
            throw createError(retrieverName, 'fetchPiecesFromProviders', `All providers failed to serve piece ${pieceCid.toString()}. Details: ${failureDetails}`);
        }
        throw error;
    }
}
