// @filoz/synapse-core/dist/src/warm-storage/add-approved-provider.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
export async function addApprovedProvider(client, options) {
    const { request } = await simulateContract(client, addApprovedProviderCall({
        chain: client.chain,
        providerId: options.providerId,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function addApprovedProviderSync(client, options) {
    const hash = await addApprovedProvider(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractAddApprovedProviderEvent(receipt.logs);
    return { receipt, event };
}
export function addApprovedProviderCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.fwss.abi,
        address: options.contractAddress ?? chain.contracts.fwss.address,
        functionName: 'addApprovedProvider',
        args: [options.providerId],
    };
}
export function extractAddApprovedProviderEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.fwss,
        logs,
        eventName: 'ProviderApproved',
        strict: true,
    });
    if (!log)
        throw new Error('`ProviderApproved` event not found.');
    return log;
}
