// @filoz/synapse-core/dist/src/warm-storage/get-pdp-data-set.js
import { isAddressEqual } from 'viem';
import { multicall } from 'viem/actions';
import { asChain } from "../chains.js";
import { dataSetLiveCall } from "../pdp-verifier/data-set-live.js";
import { getDataSetListenerCall } from "../pdp-verifier/get-data-set-listener.js";
import { getPDPProviderCall, parsePDPProvider } from "../sp-registry/get-pdp-provider.js";
import { getAllDataSetMetadataCall, parseAllDataSetMetadata } from "./get-all-data-set-metadata.js";
import { getDataSet } from "./get-data-set.js";
export async function getPdpDataSet(client, options) {
    const data = await getDataSet(client, options);
    if (!data) {
        return undefined;
    }
    const pdpInfo = await readPdpDataSetInfo(client, {
        dataSetInfo: data,
        providerId: data.providerId,
    });
    return {
        ...data,
        ...pdpInfo,
    };
}
export async function readPdpDataSetInfo(client, options) {
    const chain = asChain(client.chain);
    const [live, listener, _metadata, _pdpProvider] = await multicall(client, {
        allowFailure: false,
        contracts: [
            dataSetLiveCall({
                chain: client.chain,
                dataSetId: options.dataSetInfo.dataSetId,
            }),
            getDataSetListenerCall({
                chain: client.chain,
                dataSetId: options.dataSetInfo.dataSetId,
            }),
            getAllDataSetMetadataCall({
                chain: client.chain,
                dataSetId: options.dataSetInfo.dataSetId,
            }),
            getPDPProviderCall({
                chain: client.chain,
                providerId: options.providerId,
            }),
        ],
    });
    const pdpProvider = parsePDPProvider(_pdpProvider);
    const metadata = parseAllDataSetMetadata(_metadata);
    return {
        live,
        managed: isAddressEqual(listener, chain.contracts.fwss.address),
        cdn: options.dataSetInfo.cdnRailId > 0n && 'withCDN' in metadata,
        metadata,
        provider: pdpProvider,
    };
}
