// @filoz/synapse-core/dist/src/warm-storage/read-addresses.js
import { multicall } from 'viem/actions';
import { asChain } from "../chains.js";
export async function readAddresses(client, options = {}) {
    const chain = asChain(client.chain);
    const contractAddress = options.contractAddress ?? chain.contracts.fwss.address;
    const addresses = await multicall(client, {
        allowFailure: false,
        contracts: [
            {
                address: contractAddress,
                abi: chain.contracts.fwss.abi,
                functionName: 'paymentsContractAddress',
            },
            {
                address: contractAddress,
                abi: chain.contracts.fwss.abi,
                functionName: 'viewContractAddress',
            },
            {
                address: contractAddress,
                abi: chain.contracts.fwss.abi,
                functionName: 'pdpVerifierAddress',
            },
            {
                address: contractAddress,
                abi: chain.contracts.fwss.abi,
                functionName: 'serviceProviderRegistry',
            },
            {
                address: contractAddress,
                abi: chain.contracts.fwss.abi,
                functionName: 'sessionKeyRegistry',
            },
            {
                address: contractAddress,
                abi: chain.contracts.fwss.abi,
                functionName: 'usdfcTokenAddress',
            },
            {
                address: contractAddress,
                abi: chain.contracts.fwss.abi,
                functionName: 'filBeamBeneficiaryAddress',
            },
        ],
    });
    return {
        payments: addresses[0],
        warmStorageView: addresses[1],
        pdpVerifier: addresses[2],
        serviceProviderRegistry: addresses[3],
        sessionKeyRegistry: addresses[4],
        usdfcToken: addresses[5],
        filBeamBeneficiary: addresses[6],
    };
}
