// @filoz/synapse-core/dist/src/warm-storage/remove-approved-provider.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
export async function removeApprovedProvider(client, options) {
    const { request } = await simulateContract(client, removeApprovedProviderCall({
        chain: client.chain,
        providerId: options.providerId,
        index: options.index,
        contractAddress: options.contractAddress,
    }));
    const hash = await writeContract(client, request);
    return hash;
}
export async function removeApprovedProviderSync(client, options) {
    const hash = await removeApprovedProvider(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractRemoveApprovedProviderEvent(receipt.logs);
    return { receipt, event };
}
export function removeApprovedProviderCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.fwss.abi,
        address: options.contractAddress ?? chain.contracts.fwss.address,
        functionName: 'removeApprovedProvider',
        args: [options.providerId, options.index],
    };
}
export function extractRemoveApprovedProviderEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.fwss,
        logs,
        eventName: 'ProviderUnapproved',
        strict: true,
    });
    if (!log)
        throw new Error('`ProviderUnapproved` event not found.');
    return log;
}
