// @filoz/synapse-core/dist/src/warm-storage/terminate-service.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
export async function terminateService(client, options) {
    const { request } = await simulateContract(client, terminateServiceCall({
        chain: client.chain,
        dataSetId: options.dataSetId,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function terminateServiceSync(client, options) {
    const hash = await terminateService(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractTerminateServiceEvent(receipt.logs);
    return { receipt, event };
}
export function terminateServiceCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.fwss.abi,
        address: options.contractAddress ?? chain.contracts.fwss.address,
        functionName: 'terminateService',
        args: [options.dataSetId],
    };
}
export function extractTerminateServiceEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.fwss,
        logs,
        eventName: 'ServiceTerminated',
        strict: true,
    });
    if (!log)
        throw new Error('`ServiceTerminated` event not found.');
    return log;
}
