// @filoz/synapse-core/dist/src/utils/decode-pdp-errors.js
import { AbiErrorSignatureNotFoundError, decodeErrorResult } from 'viem';
import { formatAbiItem, formatAbiItemWithArgs } from 'viem/utils';
import * as Abis from "../abis/index.js";
export function decodePDPError(error) {
    const regex = /(?:vm error|revert reason)=\[(.*?)\]/g;
    const matches = error.matchAll(regex);
    for (const match of matches) {
        const extractedContent = match[1];
        if (extractedContent?.startsWith('0x')) {
            let error;
            try {
                const value = decodeErrorResult({
                    abi: Abis.fwss,
                    data: extractedContent,
                });
                return `Warm Storage\n${formatPDPError(value)}`;
            }
            catch (err) {
                error = err;
            }
            if (error instanceof AbiErrorSignatureNotFoundError) {
                try {
                    const value = decodeErrorResult({
                        abi: Abis.filecoinPay,
                        data: extractedContent,
                    });
                    return `Payments\n${formatPDPError(value)}`;
                }
                catch (err) {
                    error = err;
                }
            }
            if (error instanceof AbiErrorSignatureNotFoundError) {
                try {
                    const value = decodeErrorResult({
                        abi: Abis.pdp,
                        data: extractedContent,
                    });
                    return `PDP Verifier\n${formatPDPError(value)}`;
                }
                catch (err) {
                    error = err;
                }
            }
            return `Unable to decode error\n${error}`;
        }
        else if (extractedContent?.startsWith('Error(')) {
            return `\n${extractedContent.replace('Error(', '').replace(')', '')}`;
        }
    }
    return `Service Provider PDP\n${error}`;
}
function formatPDPError(error) {
    const errorWithParams = error.abiItem
        ? formatAbiItem(error.abiItem, {
            includeName: true,
        })
        : undefined;
    const formattedArgs = error.args
        ? formatAbiItemWithArgs({
            abiItem: error.abiItem,
            args: error.args,
            includeName: false,
            includeFunctionName: false,
        })
        : undefined;
    return [
        errorWithParams ? `${errorWithParams}` : '',
        formattedArgs && formattedArgs !== '()'
            ? `${[...Array(error.errorName?.length ?? 0).keys()].map(() => ' ').join('')}${formattedArgs}`
            : '',
    ].join('\n');
}
