// @filoz/synapse-core/dist/src/utils/epoch.js
import { TIME_CONSTANTS } from "./constants.js";
export function epochToDate(epoch, genesisTimestamp) {
    const epochDuration = TIME_CONSTANTS.EPOCH_DURATION;
    const timestampSeconds = genesisTimestamp + epoch * epochDuration;
    return new Date(timestampSeconds * 1000);
}
export function dateToEpoch(date, genesisTimestamp) {
    const epochDuration = TIME_CONSTANTS.EPOCH_DURATION;
    const timestampSeconds = Math.floor(date.getTime() / 1000);
    const secondsSinceGenesis = timestampSeconds - genesisTimestamp;
    return Math.floor(secondsSinceGenesis / epochDuration);
}
export function timeUntilEpoch(futureEpoch, currentEpoch) {
    const epochDifference = futureEpoch - currentEpoch;
    const seconds = epochDifference * TIME_CONSTANTS.EPOCH_DURATION;
    return {
        epochs: epochDifference,
        seconds,
        minutes: seconds / 60,
        hours: seconds / 3600,
        days: seconds / 86400,
    };
}
export function calculateLastProofDate(nextChallengeEpoch, maxProvingPeriod, genesisTimestamp) {
    if (nextChallengeEpoch === 0) {
        return null;
    }
    const lastProofEpoch = nextChallengeEpoch - maxProvingPeriod;
    if (lastProofEpoch <= 0) {
        return null;
    }
    return epochToDate(lastProofEpoch, genesisTimestamp);
}
