// @filoz/synapse-core/dist/src/utils/metadata.js
export const METADATA_LIMITS = {
    MAX_KEY_LENGTH: 32,
    MAX_VALUE_LENGTH: 128,
    MAX_KEYS_PER_DATASET: 10,
    MAX_KEYS_PER_PIECE: 5,
};
export function metadataArrayToObject(metadataArray) {
    const [keys, values] = metadataArray;
    const metadata = {};
    for (let i = 0; i < keys.length; i++) {
        metadata[keys[i]] = values[i];
    }
    return metadata;
}
export function datasetMetadataObjectToEntry(metadataObject, metadataInternal) {
    const obj = {
        ...(metadataObject ?? {}),
        ...(metadataInternal?.cdn ? { withCDN: '' } : {}),
    };
    const entries = Object.entries(obj)
        .sort((a, b) => a[0].localeCompare(b[0]))
        .map(([key, value]) => ({ key, value }));
    if (entries.length > METADATA_LIMITS.MAX_KEYS_PER_DATASET) {
        throw new Error('Metadata exceeds the maximum number of keys per data set');
    }
    for (const entry of entries) {
        if (entry.key.length > METADATA_LIMITS.MAX_KEY_LENGTH) {
            throw new Error('Metadata key exceeds the maximum length');
        }
        if (entry.value.length > METADATA_LIMITS.MAX_VALUE_LENGTH) {
            throw new Error('Metadata value exceeds the maximum length');
        }
    }
    return entries;
}
export function pieceMetadataObjectToEntry(metadataObject, metadataInternal) {
    const obj = {
        ...(metadataObject ?? {}),
        ...(metadataInternal?.ipni ? { withIPNI: '' } : {}),
        ...(metadataInternal?.ipfsRootCID ? { ipfsRootCID: metadataInternal.ipfsRootCID } : {}),
    };
    const entries = Object.entries(obj)
        .sort((a, b) => a[0].localeCompare(b[0]))
        .map(([key, value]) => ({ key, value }));
    if (entries.length > METADATA_LIMITS.MAX_KEYS_PER_PIECE) {
        throw new Error('Metadata exceeds the maximum number of keys per piece');
    }
    for (const entry of entries) {
        if (entry.key.length > METADATA_LIMITS.MAX_KEY_LENGTH) {
            throw new Error('Metadata key exceeds the maximum length');
        }
        if (entry.value.length > METADATA_LIMITS.MAX_VALUE_LENGTH) {
            throw new Error('Metadata value exceeds the maximum length');
        }
    }
    return entries;
}
