// @filoz/synapse-core/dist/src/utils/pdp-capabilities.js
import { base58btc } from 'multiformats/bases/base58';
import { bytesToHex, fromHex, hexToString, isHex, numberToBytes, stringToHex, toBytes } from 'viem';
import { z } from 'zod';
import { ZodValidationError } from "../errors/base.js";
import { capabilitiesListToObject, decodeAddressCapability } from "./capabilities.js";
import { zHex } from "./schemas.js";
export const PDPOfferingSchema = z.object({
    serviceURL: zHex,
    minPieceSizeInBytes: zHex,
    maxPieceSizeInBytes: zHex,
    storagePricePerTibPerDay: zHex,
    minProvingPeriodInEpochs: zHex,
    location: zHex,
    paymentTokenAddress: zHex,
    ipniPiece: zHex.optional(),
    ipniIpfs: zHex.optional(),
    ipniPeerId: zHex.optional(),
});
export const CAP_SERVICE_URL = 'serviceURL';
export const CAP_MIN_PIECE_SIZE = 'minPieceSizeInBytes';
export const CAP_MAX_PIECE_SIZE = 'maxPieceSizeInBytes';
export const CAP_STORAGE_PRICE = 'storagePricePerTibPerDay';
export const CAP_MIN_PROVING_PERIOD = 'minProvingPeriodInEpochs';
export const CAP_LOCATION = 'location';
export const CAP_PAYMENT_TOKEN = 'paymentTokenAddress';
export const CAP_IPNI_PIECE = 'ipniPiece';
export const CAP_IPNI_IPFS = 'ipniIpfs';
export const CAP_IPNI_PEER_ID = 'ipniPeerId';
export const CAP_IPNI_PEER_ID_LEGACY = 'IPNIPeerID';
export function decodePDPOffering(provider) {
    const capabilities = capabilitiesListToObject(provider.product.capabilityKeys, provider.productCapabilityValues);
    const parsed = PDPOfferingSchema.safeParse(capabilities);
    if (!parsed.success) {
        throw new ZodValidationError(parsed.error);
    }
    return decodePDPCapabilities(parsed.data);
}
export function decodePDPCapabilities(capabilities) {
    const required = {
        serviceURL: hexToString(capabilities.serviceURL),
        minPieceSizeInBytes: BigInt(capabilities.minPieceSizeInBytes),
        maxPieceSizeInBytes: BigInt(capabilities.maxPieceSizeInBytes),
        storagePricePerTibPerDay: BigInt(capabilities.storagePricePerTibPerDay),
        minProvingPeriodInEpochs: BigInt(capabilities.minProvingPeriodInEpochs),
        location: hexToString(capabilities.location),
        paymentTokenAddress: decodeAddressCapability(capabilities.paymentTokenAddress),
    };
    const optional = {
        ipniPiece: CAP_IPNI_PIECE in capabilities ? capabilities[CAP_IPNI_PIECE] === '0x01' : false,
        ipniIpfs: CAP_IPNI_IPFS in capabilities ? capabilities[CAP_IPNI_IPFS] === '0x01' : false,
        ipniPeerId: CAP_IPNI_PEER_ID in capabilities
            ? base58btc.encode(fromHex(capabilities[CAP_IPNI_PEER_ID], 'bytes'))
            : CAP_IPNI_PEER_ID_LEGACY in capabilities
                ? base58btc.encode(fromHex(capabilities[CAP_IPNI_PEER_ID_LEGACY], 'bytes'))
                : undefined,
    };
    return { ...required, ...optional };
}
export function encodePDPCapabilities(pdpOffering, capabilities) {
    const capabilityKeys = [];
    const capabilityValues = [];
    capabilityKeys.push(CAP_SERVICE_URL);
    capabilityValues.push(stringToHex(pdpOffering.serviceURL));
    capabilityKeys.push(CAP_MIN_PIECE_SIZE);
    capabilityValues.push(bytesToHex(numberToBytes(pdpOffering.minPieceSizeInBytes)));
    capabilityKeys.push(CAP_MAX_PIECE_SIZE);
    capabilityValues.push(bytesToHex(numberToBytes(pdpOffering.maxPieceSizeInBytes)));
    if (pdpOffering.ipniPiece) {
        capabilityKeys.push(CAP_IPNI_PIECE);
        capabilityValues.push('0x01');
    }
    if (pdpOffering.ipniIpfs) {
        capabilityKeys.push(CAP_IPNI_IPFS);
        capabilityValues.push('0x01');
    }
    capabilityKeys.push(CAP_STORAGE_PRICE);
    capabilityValues.push(bytesToHex(numberToBytes(pdpOffering.storagePricePerTibPerDay)));
    capabilityKeys.push(CAP_MIN_PROVING_PERIOD);
    capabilityValues.push(bytesToHex(numberToBytes(pdpOffering.minProvingPeriodInEpochs)));
    capabilityKeys.push(CAP_LOCATION);
    capabilityValues.push(stringToHex(pdpOffering.location));
    capabilityKeys.push(CAP_PAYMENT_TOKEN);
    capabilityValues.push(pdpOffering.paymentTokenAddress);
    if (capabilities != null) {
        for (const [key, value] of Object.entries(capabilities)) {
            capabilityKeys.push(key);
            if (!value) {
                capabilityValues.push('0x01');
            }
            else if (isHex(value)) {
                capabilityValues.push(value);
            }
            else {
                capabilityValues.push(bytesToHex(toBytes(value)));
            }
        }
    }
    return [capabilityKeys, capabilityValues];
}
