// @filoz/synapse-core/dist/src/utils/streams.js
export function isReadableStream(value) {
    return (typeof value === 'object' &&
        value !== null &&
        'getReader' in value &&
        typeof value.getReader === 'function');
}
export function asReadableStream(data) {
    return isReadableStream(data) ? data : asyncIterableToReadableStream(data);
}
export function isAsyncIterable(value) {
    return (typeof value === 'object' &&
        value !== null &&
        Symbol.asyncIterator in value &&
        typeof value[Symbol.asyncIterator] === 'function');
}
export function asyncIterableToReadableStream(iterable) {
    if (!isAsyncIterable(iterable)) {
        throw new Error('Input must be an AsyncIterable');
    }
    if (typeof ReadableStream.from === 'function') {
        return ReadableStream.from(iterable);
    }
    const iterator = iterable[Symbol.asyncIterator]();
    return new ReadableStream({
        async pull(controller) {
            try {
                const { value, done } = await iterator.next();
                if (done) {
                    controller.close();
                }
                else {
                    controller.enqueue(value);
                }
            }
            catch (error) {
                if (iterator.return) {
                    try {
                        await iterator.return();
                    }
                    catch {
                    }
                }
                controller.error(error);
            }
        },
        async cancel() {
            if (iterator.return) {
                await iterator.return();
            }
        },
    });
}
export async function* uint8ArrayToAsyncIterable(data, chunkSize = 2048) {
    for (let i = 0; i < data.length; i += chunkSize) {
        yield data.subarray(i, i + chunkSize);
    }
}
export function isUint8Array(value) {
    return value instanceof Uint8Array || (ArrayBuffer.isView(value) && value.constructor.name === 'Uint8Array');
}
