// @filoz/synapse-core/dist/src/mocks/jsonrpc/service-registry.js
import { decodeFunctionData, encodeAbiParameters, isAddressEqual } from 'viem';
import * as Abis from "../../abis/index.js";
import { encodePDPCapabilities } from "../../utils/pdp-capabilities.js";
const EMPTY_PROVIDER_INFO = {
    serviceProvider: '0x0000000000000000000000000000000000000000',
    payee: '0x0000000000000000000000000000000000000000',
    name: '',
    description: '',
    isActive: false,
};
const EMPTY_PROVIDER_INFO_VIEW = {
    providerId: 0n,
    info: EMPTY_PROVIDER_INFO,
};
const _EMPTY_PROVIDER_WITH_PRODUCT = [
    {
        providerId: 0n,
        providerInfo: EMPTY_PROVIDER_INFO,
        product: {
            productType: 0,
            capabilityKeys: [],
            isActive: false,
        },
        productCapabilityValues: [],
    },
];
export function mockServiceProviderRegistry(providers) {
    const activeProviders = providers.filter((p) => p.providerInfo.isActive);
    return {
        getProvider: ([providerId]) => {
            const provider = providers.find((p) => p.providerId === providerId);
            if (!provider) {
                throw new Error('Provider not found');
            }
            return [
                {
                    providerId,
                    info: provider.providerInfo,
                },
            ];
        },
        getAllActiveProviders: ([offset, limit]) => {
            const providerIds = activeProviders.map((p) => p.providerId).slice(Number(offset), Number(offset + limit));
            const hasMore = offset + limit < activeProviders.length;
            return [providerIds, hasMore];
        },
        getProviderCount: () => {
            return [BigInt(providers.length)];
        },
        activeProviderCount: () => {
            return [BigInt(activeProviders.length)];
        },
        isProviderActive: ([providerId]) => {
            const provider = providers.find((p) => p.providerId === providerId);
            return [provider?.providerInfo.isActive ?? false];
        },
        isRegisteredProvider: ([address]) => {
            const provider = providers.find((p) => isAddressEqual(address, p.providerInfo.serviceProvider));
            return [provider != null];
        },
        REGISTRATION_FEE: () => {
            return 0n;
        },
        getProviderWithProduct: ([providerId, productType]) => {
            const provider = providers.find((p) => p.providerId === providerId);
            if (!provider) {
                throw new Error('Provider not found');
            }
            const product = provider.products.find((p) => p?.productType === productType);
            if (!product) {
                throw new Error('Service does not exist');
            }
            const [capabilityKeys, productCapabilityValues] = encodePDPCapabilities(product.offering);
            return [
                {
                    providerId,
                    providerInfo: provider.providerInfo,
                    product: {
                        productType: product.productType,
                        capabilityKeys,
                        isActive: product.isActive,
                    },
                    productCapabilityValues,
                },
            ];
        },
        getProvidersByProductType: ([productType, onlyActive, offset, limit]) => {
            if (!providers) {
                return [
                    {
                        providers: [],
                        hasMore: false,
                    },
                ];
            }
            const filteredProviders = [];
            for (let i = 0; i < providers.length; i++) {
                const providerInfoView = providers[i];
                const providerId = providerInfoView.providerId;
                const providerInfo = providers[i].providerInfo;
                if (onlyActive && !providerInfo.isActive) {
                    continue;
                }
                const product = providers[i].products.find((p) => p?.productType === productType && p?.isActive);
                if (!product) {
                    continue;
                }
                const [capabilityKeys, productCapabilityValues] = encodePDPCapabilities(product.offering);
                filteredProviders.push({
                    providerId,
                    providerInfo,
                    product: {
                        productType: 0,
                        capabilityKeys,
                        isActive: product.isActive,
                    },
                    productCapabilityValues,
                });
            }
            const hasMore = Number(offset + limit) < filteredProviders.length;
            return [
                {
                    providers: filteredProviders.slice(Number(offset), Number(offset + limit)),
                    hasMore,
                },
            ];
        },
        getProviderByAddress: ([address]) => {
            for (const provider of providers) {
                if (address === provider.providerInfo.serviceProvider) {
                    return [
                        {
                            providerId: provider.providerId,
                            info: provider.providerInfo,
                        },
                    ];
                }
            }
            return [EMPTY_PROVIDER_INFO_VIEW];
        },
        getProviderIdByAddress: ([address]) => {
            for (const provider of providers) {
                if (address === provider.providerInfo.serviceProvider) {
                    return [provider.providerId];
                }
            }
            return [0n];
        },
    };
}
export function serviceProviderRegistryCallHandler(data, value, from, options) {
    const { functionName, args } = decodeFunctionData({
        abi: Abis.serviceProviderRegistry,
        data: data,
    });
    if (options.debug) {
        console.debug('Service Provider Registry: calling function', functionName, 'with args', args);
    }
    switch (functionName) {
        case 'getProviderByAddress': {
            if (!options.serviceRegistry?.getProviderByAddress) {
                throw new Error('Service Provider Registry: getProviderByAddress is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'getProviderByAddress')
                .outputs, options.serviceRegistry.getProviderByAddress(args));
        }
        case 'getProviderIdByAddress': {
            if (!options.serviceRegistry?.getProviderIdByAddress) {
                throw new Error('Service Provider Registry: getProviderIdByAddress is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'getProviderIdByAddress')
                .outputs, options.serviceRegistry.getProviderIdByAddress(args));
        }
        case 'getProvider': {
            if (!options.serviceRegistry?.getProvider) {
                throw new Error('Service Provider Registry: getProvider is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'getProvider').outputs, options.serviceRegistry.getProvider(args));
        }
        case 'getProviderWithProduct': {
            if (!options.serviceRegistry?.getProviderWithProduct) {
                throw new Error('Service Provider Registry: getProviderWithProduct is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'getProviderWithProduct')
                .outputs, options.serviceRegistry.getProviderWithProduct(args));
        }
        case 'getProvidersByProductType': {
            if (!options.serviceRegistry?.getProvidersByProductType) {
                throw new Error('Service Provider Registry: getProvidersByProductType is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'getProvidersByProductType')
                .outputs, options.serviceRegistry.getProvidersByProductType(args));
        }
        case 'getAllActiveProviders': {
            if (!options.serviceRegistry?.getAllActiveProviders) {
                throw new Error('Service Provider Registry: getAllActiveProviders is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'getAllActiveProviders')
                .outputs, options.serviceRegistry.getAllActiveProviders(args));
        }
        case 'getProviderCount': {
            if (!options.serviceRegistry?.getProviderCount) {
                throw new Error('Service Provider Registry: getProviderCount is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'getProviderCount').outputs, options.serviceRegistry.getProviderCount(args));
        }
        case 'activeProviderCount': {
            if (!options.serviceRegistry?.activeProviderCount) {
                throw new Error('Service Provider Registry: activeProviderCount is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'activeProviderCount')
                .outputs, options.serviceRegistry.activeProviderCount(args));
        }
        case 'isProviderActive': {
            if (!options.serviceRegistry?.isProviderActive) {
                throw new Error('Service Provider Registry: isProviderActive is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'isProviderActive').outputs, options.serviceRegistry.isProviderActive(args));
        }
        case 'isRegisteredProvider': {
            if (!options.serviceRegistry?.isRegisteredProvider) {
                throw new Error('Service Provider Registry: isRegisteredProvider is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'isRegisteredProvider')
                .outputs, options.serviceRegistry.isRegisteredProvider(args));
        }
        case 'REGISTRATION_FEE': {
            if (!options.serviceRegistry?.REGISTRATION_FEE) {
                throw new Error('Service Provider Registry: REGISTRATION_FEE is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'REGISTRATION_FEE').outputs, [options.serviceRegistry.REGISTRATION_FEE()]);
        }
        case 'registerProvider': {
            if (!options.serviceRegistry?.registerProvider) {
                throw new Error('Service Provider Registry: registerProvider is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'registerProvider').outputs, options.serviceRegistry.registerProvider(args, value, from));
        }
        case 'updateProviderInfo': {
            if (!options.serviceRegistry?.updateProviderInfo) {
                throw new Error('Service Provider Registry: updateProviderInfo is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'updateProviderInfo')
                .outputs, options.serviceRegistry.updateProviderInfo(args, value, from));
        }
        case 'removeProvider': {
            if (!options.serviceRegistry?.removeProvider) {
                throw new Error('Service Provider Registry: removeProvider is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'removeProvider').outputs, options.serviceRegistry.removeProvider(args, value, from));
        }
        case 'addProduct': {
            if (!options.serviceRegistry?.addProduct) {
                throw new Error('Service Provider Registry: addProduct is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'addProduct').outputs, options.serviceRegistry.addProduct(args, value, from));
        }
        case 'updateProduct': {
            if (!options.serviceRegistry?.updateProduct) {
                throw new Error('Service Provider Registry: updateProduct is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'updateProduct').outputs, options.serviceRegistry.updateProduct(args, value, from));
        }
        case 'removeProduct': {
            if (!options.serviceRegistry?.removeProduct) {
                throw new Error('Service Provider Registry: removeProduct is not defined');
            }
            return encodeAbiParameters(Abis.serviceProviderRegistry.find((abi) => abi.type === 'function' && abi.name === 'removeProduct').outputs, options.serviceRegistry.removeProduct(args, value, from));
        }
        default: {
            throw new Error(`Service Provider Registry: unknown function: ${functionName} with args: ${args}`);
        }
    }
}
