// @filoz/synapse-core/dist/src/sp/data-sets.js
import { asChain, getChain } from "../chains.js";
import { signCreateDataSet } from "../typed-data/sign-create-dataset.js";
import { signCreateDataSetAndAddPieces } from "../typed-data/sign-create-dataset-add-pieces.js";
import { datasetMetadataObjectToEntry, pieceMetadataObjectToEntry } from "../utils/metadata.js";
import * as SP from "./sp.js";
export async function createDataSet(client, options) {
    const chain = getChain(client.chain.id);
    const extraData = await signCreateDataSet(client, {
        clientDataSetId: options.clientDataSetId,
        payee: options.payee,
        payer: options.payer,
        metadata: datasetMetadataObjectToEntry(options.metadata, {
            cdn: options.cdn,
        }),
    });
    return SP.createDataSet({
        serviceURL: options.serviceURL,
        recordKeeper: options.recordKeeper ?? chain.contracts.fwss.address,
        extraData,
    });
}
export async function createDataSetAndAddPieces(client, options) {
    const chain = asChain(client.chain);
    const { txHash, statusUrl } = await SP.createDataSetAndAddPieces({
        serviceURL: options.serviceURL,
        recordKeeper: options.recordKeeper ?? chain.contracts.fwss.address,
        extraData: await signCreateDataSetAndAddPieces(client, {
            clientDataSetId: options.clientDataSetId,
            payee: options.payee,
            payer: options.payer,
            metadata: datasetMetadataObjectToEntry(options.metadata, {
                cdn: options.cdn,
            }),
            pieces: options.pieces.map((piece) => ({
                pieceCid: piece.pieceCid,
                metadata: pieceMetadataObjectToEntry(piece.metadata),
            })),
        }),
        pieces: options.pieces.map((piece) => piece.pieceCid),
    });
    return {
        txHash,
        statusUrl,
    };
}
