// @filoz/synapse-core/dist/src/sp/sp.js
import { HttpError, request, TimeoutError } from 'iso-web/http';
import { isHex } from 'viem';
import { AddPiecesError, CreateDataSetError, DeletePieceError, DownloadPieceError, FindPieceError, InvalidUploadSizeError, LocationHeaderError, PostPieceError, UploadPieceError, } from "../errors/pdp.js";
import * as Piece from "../piece.js";
import { RETRY_CONSTANTS, SIZE_CONSTANTS } from "../utils/constants.js";
import { createPieceUrlPDP } from "../utils/piece-url.js";
import { isUint8Array } from "../utils/streams.js";
export async function createDataSet(options) {
    const response = await request.post(new URL(`pdp/data-sets`, options.serviceURL), {
        body: JSON.stringify({
            recordKeeper: options.recordKeeper,
            extraData: options.extraData,
        }),
        headers: {
            'Content-Type': 'application/json',
        },
        timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
    });
    if (response.error) {
        if (HttpError.is(response.error)) {
            throw new CreateDataSetError(await response.error.response.text());
        }
        throw response.error;
    }
    const location = response.result.headers.get('Location');
    const hash = location?.split('/').pop();
    if (!location || !hash || !isHex(hash)) {
        throw new LocationHeaderError(location);
    }
    return {
        txHash: hash,
        statusUrl: new URL(location, options.serviceURL).toString(),
    };
}
export async function createDataSetAndAddPieces(options) {
    const response = await request.post(new URL(`pdp/data-sets/create-and-add`, options.serviceURL), {
        body: JSON.stringify({
            recordKeeper: options.recordKeeper,
            extraData: options.extraData,
            pieces: options.pieces.map((piece) => ({
                pieceCid: piece.toString(),
                subPieces: [{ subPieceCid: piece.toString() }],
            })),
        }),
        headers: {
            'Content-Type': 'application/json',
        },
        timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
    });
    if (response.error) {
        if (HttpError.is(response.error)) {
            throw new CreateDataSetError(await response.error.response.text());
        }
        throw response.error;
    }
    const location = response.result.headers.get('Location');
    const hash = location?.split('/').pop();
    if (!location || !hash || !isHex(hash)) {
        throw new LocationHeaderError(location);
    }
    return {
        txHash: hash,
        statusUrl: new URL(location, options.serviceURL).toString(),
    };
}
export async function uploadPiece(options) {
    const size = options.data.length;
    if (size < SIZE_CONSTANTS.MIN_UPLOAD_SIZE || size > SIZE_CONSTANTS.MAX_UPLOAD_SIZE) {
        throw new InvalidUploadSizeError(size);
    }
    const pieceCid = options.pieceCid;
    if (!Piece.isPieceCID(pieceCid)) {
        throw new Error(`Invalid PieceCID: ${String(options.pieceCid)}`);
    }
    const response = await request.post(new URL(`pdp/piece`, options.serviceURL), {
        body: JSON.stringify({
            pieceCid: pieceCid.toString(),
        }),
        headers: {
            'Content-Type': 'application/json',
        },
        timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
    });
    if (response.error) {
        if (HttpError.is(response.error)) {
            throw new PostPieceError(await response.error.response.text());
        }
        throw response.error;
    }
    if (response.result.status === 200) {
        return;
    }
    const location = response.result.headers.get('Location');
    const uploadUuid = location?.split('/').pop();
    if (!location || !uploadUuid) {
        throw new LocationHeaderError(location);
    }
    const uploadResponse = await request.put(new URL(`pdp/piece/upload/${uploadUuid}`, options.serviceURL), {
        body: options.data,
        headers: {
            'Content-Type': 'application/octet-stream',
            'Content-Length': options.data.length.toString(),
        },
        timeout: false,
    });
    if (uploadResponse.error) {
        if (HttpError.is(uploadResponse.error)) {
            throw new UploadPieceError(await uploadResponse.error.response.text());
        }
        throw uploadResponse.error;
    }
}
export async function uploadPieceStreaming(options) {
    const createResponse = await request.post(new URL('pdp/piece/uploads', options.serviceURL), {
        timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
        signal: options.signal,
    });
    if (createResponse.error) {
        if (HttpError.is(createResponse.error)) {
            throw new PostPieceError(`Failed to create upload session: ${await createResponse.error.response.text()}`);
        }
        throw createResponse.error;
    }
    if (createResponse.result.status !== 201) {
        throw new PostPieceError(`Expected 201 Created, got ${createResponse.result.status}`);
    }
    const location = createResponse.result.headers.get('Location');
    if (!location) {
        throw new LocationHeaderError('Upload session created but Location header missing');
    }
    const locationMatch = location.match(/\/pdp\/piece\/uploads\/([a-fA-F0-9-]+)/);
    if (!locationMatch) {
        throw new LocationHeaderError(`Invalid Location header format: ${location}`);
    }
    const uploadUuid = locationMatch[1];
    let getPieceCID = () => options.pieceCid ?? null;
    let pieceCidStream = null;
    if (options.pieceCid == null) {
        const result = Piece.createPieceCIDStream();
        pieceCidStream = result.stream;
        getPieceCID = result.getPieceCID;
    }
    const dataStream = isUint8Array(options.data)
        ? new Blob([options.data]).stream()
        : options.data;
    const size = isUint8Array(options.data) ? options.data.length : options.size;
    let bytesUploaded = 0;
    const trackingStream = new TransformStream({
        transform(chunk, controller) {
            let bytes;
            if (isUint8Array(chunk)) {
                bytes = chunk;
            }
            else if (ArrayBuffer.isView(chunk)) {
                bytes = new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength);
            }
            else {
                controller.error('Invalid chunk type only Uint8Array and TypedArray are supported');
                return;
            }
            bytesUploaded += bytes.length;
            if (bytesUploaded > SIZE_CONSTANTS.MAX_UPLOAD_SIZE) {
                controller.error(new InvalidUploadSizeError(bytesUploaded));
                return;
            }
            if (options.onProgress) {
                options.onProgress(bytesUploaded);
            }
            controller.enqueue(bytes);
        },
        flush(controller) {
            if (bytesUploaded < SIZE_CONSTANTS.MIN_UPLOAD_SIZE) {
                controller.error(new InvalidUploadSizeError(bytesUploaded));
                return;
            }
        },
    });
    const bodyStream = pieceCidStream
        ? dataStream.pipeThrough(trackingStream).pipeThrough(pieceCidStream)
        : dataStream.pipeThrough(trackingStream);
    const headers = {
        'Content-Type': 'application/octet-stream',
        ...(size != null ? { 'Content-Length': size.toString() } : {}),
    };
    const uploadResponse = await request.put(new URL(`pdp/piece/uploads/${uploadUuid}`, options.serviceURL), {
        body: bodyStream,
        headers,
        timeout: false,
        signal: options.signal,
        duplex: 'half',
    });
    if (uploadResponse.error) {
        if (HttpError.is(uploadResponse.error)) {
            throw new UploadPieceError(`Failed to upload piece: ${await uploadResponse.error.response.text()}`);
        }
        throw uploadResponse.error;
    }
    if (uploadResponse.result.status !== 204) {
        throw new UploadPieceError(`Expected 204 No Content, got ${uploadResponse.result.status}`);
    }
    const pieceCid = getPieceCID();
    if (pieceCid === null) {
        throw new Error('Failed to calculate PieceCID during upload');
    }
    const finalizeBody = JSON.stringify({
        pieceCid: pieceCid.toString(),
    });
    const finalizeResponse = await request.post(new URL(`pdp/piece/uploads/${uploadUuid}`, options.serviceURL), {
        body: finalizeBody,
        headers: {
            'Content-Type': 'application/json',
        },
        timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
        signal: options.signal,
    });
    if (finalizeResponse.error) {
        if (HttpError.is(finalizeResponse.error)) {
            throw new PostPieceError(`Failed to finalize upload: ${await finalizeResponse.error.response.text()}`);
        }
        throw finalizeResponse.error;
    }
    if (finalizeResponse.result.status !== 200) {
        throw new PostPieceError(`Expected 200 OK for finalization, got ${finalizeResponse.result.status}`);
    }
    return {
        pieceCid,
        size: bytesUploaded,
    };
}
export async function findPiece(options) {
    const { pieceCid, serviceURL } = options;
    const params = new URLSearchParams({ pieceCid: pieceCid.toString() });
    const retry = options.retry ?? false;
    const response = await request.json.get(new URL(`pdp/piece?${params.toString()}`, serviceURL), {
        signal: options.signal,
        retry: retry
            ? {
                statusCodes: [202, 404],
                retries: RETRY_CONSTANTS.RETRIES,
                factor: RETRY_CONSTANTS.FACTOR,
                minTimeout: options.pollInterval ?? 1000,
            }
            : undefined,
        timeout: options.timeout ?? RETRY_CONSTANTS.MAX_RETRY_TIME,
    });
    if (response.error) {
        if (HttpError.is(response.error)) {
            throw new FindPieceError(await response.error.response.text());
        }
        if (TimeoutError.is(response.error)) {
            throw new FindPieceError('Timeout waiting for piece to be found');
        }
        throw response.error;
    }
    const data = response.result;
    return Piece.parse(data.pieceCid);
}
export async function addPieces(options) {
    const { serviceURL, dataSetId, pieces, extraData } = options;
    const response = await request.post(new URL(`pdp/data-sets/${dataSetId}/pieces`, serviceURL), {
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            pieces: pieces.map((piece) => ({
                pieceCid: piece.toString(),
                subPieces: [{ subPieceCid: piece.toString() }],
            })),
            extraData: extraData,
        }),
        timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
    });
    if (response.error) {
        if (HttpError.is(response.error)) {
            throw new AddPiecesError(await response.error.response.text());
        }
        throw response.error;
    }
    const location = response.result.headers.get('Location');
    const txHash = location?.split('/').pop();
    if (!location || !txHash || !isHex(txHash)) {
        throw new LocationHeaderError(location);
    }
    return {
        txHash: txHash,
        statusUrl: new URL(location, serviceURL).toString(),
    };
}
export async function deletePiece(options) {
    const { serviceURL, dataSetId, pieceId, extraData } = options;
    const response = await request.json.delete(new URL(`pdp/data-sets/${dataSetId}/pieces/${pieceId}`, serviceURL), {
        body: { extraData },
        timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
    });
    if (response.error) {
        if (HttpError.is(response.error)) {
            throw new DeletePieceError(await response.error.response.text());
        }
        throw response.error;
    }
    return { hash: response.result.txHash };
}
export async function ping(serviceURL) {
    const response = await request.get(new URL(`pdp/ping`, serviceURL));
    if (response.error) {
        throw new Error('Ping failed');
    }
    return response.result;
}
export async function downloadPiece(options) {
    const url = createPieceUrlPDP({ cid: options.pieceCid.toString(), serviceURL: options.serviceURL });
    const response = await request.get(url);
    if (response.error) {
        if (HttpError.is(response.error)) {
            throw new DownloadPieceError(await response.error.response.text());
        }
        throw response.error;
    }
    return await Piece.downloadAndValidate(response.result, options.pieceCid);
}
