// @filoz/synapse-core/dist/src/sp/upload.js
import { asChain } from "../chains.js";
import { DataSetNotFoundError } from "../errors/warm-storage.js";
import * as Piece from "../piece.js";
import { signAddPieces } from "../typed-data/sign-add-pieces.js";
import { pieceMetadataObjectToEntry } from "../utils/metadata.js";
import { createPieceUrl } from "../utils/piece-url.js";
import { getPdpDataSet } from "../warm-storage/get-pdp-data-set.js";
import * as SP from "./sp.js";
export async function upload(client, options) {
    const dataSet = await getPdpDataSet(client, {
        dataSetId: options.dataSetId,
    });
    if (!dataSet) {
        throw new DataSetNotFoundError(options.dataSetId);
    }
    const chain = asChain(client.chain);
    const serviceURL = dataSet.provider.pdp.serviceURL;
    const uploadResponses = await Promise.all(options.data.map(async (file) => {
        const data = new Uint8Array(await file.arrayBuffer());
        const pieceCid = Piece.calculate(data);
        const url = createPieceUrl({
            cid: pieceCid.toString(),
            cdn: dataSet.cdn,
            address: client.account.address,
            chain: chain,
            serviceURL,
        });
        await SP.uploadPiece({
            data,
            pieceCid,
            serviceURL,
        });
        options.onEvent?.('pieceUploaded', { pieceCid, dataSet });
        await SP.findPiece({
            pieceCid,
            serviceURL,
            retry: true,
        });
        options.onEvent?.('pieceParked', { pieceCid, url, dataSet });
        return {
            pieceCid,
            url,
            metadata: { name: file.name, type: file.type },
        };
    }));
    const addPieces = await SP.addPieces({
        dataSetId: options.dataSetId,
        pieces: uploadResponses.map((response) => response.pieceCid),
        serviceURL,
        extraData: await signAddPieces(client, {
            clientDataSetId: dataSet.clientDataSetId,
            pieces: uploadResponses.map((response) => ({
                pieceCid: response.pieceCid,
                metadata: pieceMetadataObjectToEntry(response.metadata),
            })),
        }),
    });
    return { ...addPieces, pieces: uploadResponses };
}
