// @filoz/synapse-core/dist/src/sp/wait-for-create-dataset.js
import { HttpError, request } from 'iso-web/http';
import * as z from 'zod';
import { WaitForCreateDataSetError, WaitForCreateDataSetRejectedError } from "../errors/pdp.js";
import { RETRY_CONSTANTS } from "../utils/constants.js";
import { zHex, zNumberToBigInt } from "../utils/schemas.js";
export const CreateDataSetPendingSchema = z.object({
    createMessageHash: zHex,
    dataSetCreated: z.literal(false),
    service: z.string(),
    txStatus: z.union([z.literal('pending'), z.literal('confirmed')]),
    ok: z.null(),
});
export const CreateDataSetRejectedSchema = z.object({
    createMessageHash: zHex,
    dataSetCreated: z.literal(false),
    service: z.string(),
    txStatus: z.literal('rejected'),
    ok: z.literal(false),
});
export const CreateDataSetSuccessSchema = z.object({
    createMessageHash: zHex,
    dataSetCreated: z.literal(true),
    service: z.string(),
    txStatus: z.literal('confirmed'),
    ok: z.literal(true),
    dataSetId: zNumberToBigInt,
});
const schema = z.discriminatedUnion('txStatus', [CreateDataSetRejectedSchema, CreateDataSetSuccessSchema]);
export async function waitForCreateDataSet(options) {
    const response = await request.json.get(options.statusUrl, {
        async onResponse(response) {
            if (response.ok) {
                const data = (await response.clone().json());
                if (data.dataSetCreated === false) {
                    throw new Error('Still pending');
                }
            }
        },
        retry: {
            shouldRetry: (ctx) => ctx.error.message === 'Still pending',
            retries: RETRY_CONSTANTS.RETRIES,
            factor: RETRY_CONSTANTS.FACTOR,
            minTimeout: options.pollInterval ?? RETRY_CONSTANTS.DELAY_TIME,
        },
        timeout: options.timeout ?? RETRY_CONSTANTS.MAX_RETRY_TIME,
    });
    if (response.error) {
        if (HttpError.is(response.error)) {
            throw new WaitForCreateDataSetError(await response.error.response.text());
        }
        throw response.error;
    }
    const data = schema.parse(response.result);
    if (data.txStatus === 'rejected') {
        throw new WaitForCreateDataSetRejectedError(data);
    }
    return data;
}
