// @filoz/synapse-core/dist/src/sp/wait-for-add-pieces.js
import { HttpError, request } from 'iso-web/http';
import * as z from 'zod';
import { WaitForAddPiecesError, WaitForAddPiecesRejectedError } from "../errors/pdp.js";
import { RETRY_CONSTANTS } from "../utils/constants.js";
import { zHex, zNumberToBigInt } from "../utils/schemas.js";
export const AddPiecesPendingSchema = z.object({
    txHash: zHex,
    txStatus: z.literal('pending'),
    dataSetId: zNumberToBigInt,
    pieceCount: z.number(),
    addMessageOk: z.null(),
    piecesAdded: z.literal(false),
});
export const AddPiecesRejectedSchema = z.object({
    txHash: zHex,
    txStatus: z.literal('rejected'),
    dataSetId: zNumberToBigInt,
    pieceCount: z.number(),
    addMessageOk: z.null(),
    piecesAdded: z.literal(false),
});
export const AddPiecesSuccessSchema = z.object({
    txHash: zHex,
    txStatus: z.literal('confirmed'),
    dataSetId: zNumberToBigInt,
    pieceCount: z.number(),
    addMessageOk: z.literal(true),
    piecesAdded: z.literal(true),
    confirmedPieceIds: z.array(zNumberToBigInt),
});
const schema = z.discriminatedUnion('txStatus', [AddPiecesRejectedSchema, AddPiecesSuccessSchema]);
export async function waitForAddPieces(options) {
    const response = await request.json.get(options.statusUrl, {
        async onResponse(response) {
            if (response.ok) {
                const data = (await response.clone().json());
                if (data.piecesAdded === false) {
                    throw new Error('Still pending');
                }
            }
        },
        retry: {
            shouldRetry: (ctx) => ctx.error.message === 'Still pending',
            retries: RETRY_CONSTANTS.RETRIES,
            factor: RETRY_CONSTANTS.FACTOR,
            minTimeout: options.pollInterval ?? RETRY_CONSTANTS.DELAY_TIME,
        },
        timeout: options.timeout ?? RETRY_CONSTANTS.MAX_RETRY_TIME,
    });
    if (response.error) {
        if (HttpError.is(response.error)) {
            throw new WaitForAddPiecesError(await response.error.response.text());
        }
        throw response.error;
    }
    const data = schema.parse(response.result);
    if (data.txStatus === 'rejected') {
        throw new WaitForAddPiecesRejectedError(data);
    }
    return data;
}
